// @stechquick/algae/lib/helpers/clientDB/clientDB.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientDB = void 0;
var createPromiseData_1 = require("../createPromiseData");
var ClientDBUpgraderStore = /** @class */ (function () {
    function ClientDBUpgraderStore(store) {
        this.store = store;
    }
    ClientDBUpgraderStore.prototype.getIndex = function (name) {
        if (!this.store.indexNames.contains(name)) {
            return undefined;
        }
        var index = this.store.index(name);
        return { multiEntry: index.multiEntry, unique: index.unique };
    };
    ClientDBUpgraderStore.prototype.deleteIndex = function (name) {
        this.store.deleteIndex(name);
    };
    return ClientDBUpgraderStore;
}());
var ClientDBUpgrader = /** @class */ (function () {
    function ClientDBUpgrader(db, tx) {
        this.db = db;
        this.tx = tx;
    }
    ClientDBUpgrader.prototype.createStore = function (storeName, options) {
        var _a, _b;
        if (options === void 0) { options = {}; }
        var indexes = options.indexes;
        var keyPath = options.keyPath instanceof Array ? options.keyPath.map(function (keyPath) { return keyPath.toString(); }) : (_a = options.keyPath) === null || _a === void 0 ? void 0 : _a.toString();
        function createTable(_a) {
            var reCreate = _a.reCreate, db = _a.db;
            if (reCreate) {
                if (!options.reCreatePermittedIfKeyMismatch) {
                    var keys = { old: store.keyPath, new: keyPath };
                    throw new Error("Store " + storeName + " reCreate is not permitted for key mismatch. keys: " + JSON.stringify(keys));
                }
                db.deleteObjectStore(storeName);
            }
            return db.createObjectStore(storeName, { keyPath: keyPath });
        }
        var store = (_b = this.getStorePriv(storeName)) !== null && _b !== void 0 ? _b : createTable({ db: this.db, reCreate: false });
        if (keyPath) {
            if (keyPath instanceof Array != store.keyPath instanceof Array) { // type mismatch
                store = createTable({ db: this.db, reCreate: true });
            }
            else if (keyPath.length != store.keyPath.length) { // (array | string).length mismatch
                store = createTable({ db: this.db, reCreate: true });
            }
            else if (keyPath instanceof Array) { // both array
                var keyMismatch = keyPath.find(function (curKeyPath, index) { return store.keyPath[index] != curKeyPath; });
                if (keyMismatch) {
                    store = createTable({ db: this.db, reCreate: true });
                }
            }
            else { // both string
                if (keyPath != store.keyPath) {
                    store = createTable({ db: this.db, reCreate: true });
                }
            }
        }
        function deleteIndex(name) { store.deleteIndex(name); }
        function createIndex(name, options) { store.createIndex(name, name, options); }
        function recreateIndex(oldIndex, newIndex) {
            if (newIndex.unique == oldIndex.unique && oldIndex.multiEntry == newIndex.multiEntry) {
                return;
            }
            deleteIndex(oldIndex.name);
            createIndex(oldIndex.name, newIndex);
        }
        indexes && Object.keys(indexes).forEach(function (indexName) {
            var newIndex = indexes[indexName]; // ! -> iterating indexes, obviously indexName is a key
            if (!store.indexNames.contains(indexName)) {
                createIndex(indexName, newIndex);
                return;
            }
            var oldIndex = store.index(indexName);
            recreateIndex(oldIndex, newIndex);
        });
        Array.from(store.indexNames).forEach(function (indexName) {
            if (!indexes) {
                deleteIndex(indexName);
                return;
            }
            var newIndex = indexes[indexName];
            if (!newIndex) {
                deleteIndex(indexName);
                return;
            }
            var oldIndex = store.index(indexName);
            recreateIndex(oldIndex, newIndex);
        });
    };
    ClientDBUpgrader.prototype.deleteStore = function (storeName) {
        this.db.deleteObjectStore(storeName);
    };
    ClientDBUpgrader.prototype.storeExists = function (storeName) {
        return this.getStore(storeName) != undefined;
    };
    ClientDBUpgrader.prototype.getStorePriv = function (storeName) {
        if (!this.db.objectStoreNames.contains(storeName)) {
            return undefined;
        }
        return this.tx.objectStore(storeName);
    };
    ClientDBUpgrader.prototype.getStore = function (storeName) {
        var store = this.getStorePriv(storeName);
        if (!store) {
            return undefined;
        }
        ;
        return new ClientDBUpgraderStore(store);
    };
    return ClientDBUpgrader;
}());
var ClientDBStoreReadOnly = /** @class */ (function () {
    function ClientDBStoreReadOnly(store) {
        this.store = store;
    }
    ClientDBStoreReadOnly.prototype.allRecords = function () {
        return this.queryPrivate(undefined, { type: "multi" });
    };
    ClientDBStoreReadOnly.prototype.query = function (key, options) {
        return this.queryPrivate(key, options);
    };
    ClientDBStoreReadOnly.prototype.queryPrivate = function (key, options) {
        var prom = (0, createPromiseData_1.CreatePromiseData)();
        var request = options.type == "single"
            ? this.store.get(key)
            : this.store.getAll(key, options.count);
        request.addEventListener("success", function (ev) {
            prom.resolver(ev.target.result);
        });
        request.addEventListener("error", function (ev) {
            prom.reject(ev.target.error);
        });
        return prom.promise;
    };
    return ClientDBStoreReadOnly;
}());
var ClientDBStoreReadWrite = /** @class */ (function (_super) {
    __extends(ClientDBStoreReadWrite, _super);
    function ClientDBStoreReadWrite() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ClientDBStoreReadWrite.prototype.insert = function (value, type, key) {
        return __awaiter(this, void 0, void 0, function () {
            var prom, request;
            return __generator(this, function (_a) {
                prom = (0, createPromiseData_1.CreatePromiseData)();
                request = type == "upsert" ? this.store.put(value, key) : this.store.add(value, key);
                request.addEventListener("success", function (ev) { prom.resolver(); });
                request.addEventListener("error", function (ev) { prom.reject(ev.target.error); });
                return [2 /*return*/];
            });
        });
    };
    ClientDBStoreReadWrite.prototype.delete = function (key) {
        return __awaiter(this, void 0, void 0, function () {
            var prom, request;
            return __generator(this, function (_a) {
                prom = (0, createPromiseData_1.CreatePromiseData)();
                request = this.store.delete(key);
                request.addEventListener("success", function (ev) { prom.resolver(); });
                request.addEventListener("error", function (ev) { prom.reject(ev.target.error); });
                return [2 /*return*/];
            });
        });
    };
    return ClientDBStoreReadWrite;
}(ClientDBStoreReadOnly));
var ClientDBTransactionReadOnly = /** @class */ (function () {
    function ClientDBTransactionReadOnly(tx, storeNames) {
        this.tx = tx;
        this.storeNames = storeNames;
    }
    ClientDBTransactionReadOnly.prototype.createInternaltStore = function (storeName) {
        var storeNamePriv = storeName;
        if (!storeNamePriv) {
            if (this.storeNames instanceof Array) {
                throw new Error();
            }
            storeNamePriv = this.storeNames;
        }
        return this.tx.objectStore(storeNamePriv);
    };
    ClientDBTransactionReadOnly.prototype.getStore = function (storeName) {
        return new ClientDBStoreReadOnly(this.createInternaltStore(storeName));
    };
    return ClientDBTransactionReadOnly;
}());
var ClientDBTransactionReadWrite = /** @class */ (function (_super) {
    __extends(ClientDBTransactionReadWrite, _super);
    function ClientDBTransactionReadWrite() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ClientDBTransactionReadWrite.prototype.getStore = function (storeName) {
        return new ClientDBStoreReadWrite(this.createInternaltStore(storeName));
    };
    return ClientDBTransactionReadWrite;
}(ClientDBTransactionReadOnly));
var clientDBTransactionDict = {
    readonly: ClientDBTransactionReadOnly,
    readwrite: ClientDBTransactionReadWrite,
};
var ClientDBConnection = /** @class */ (function () {
    function ClientDBConnection(db) {
        this.db = db;
    }
    ClientDBConnection.prototype.closeDB = function () {
        this.db.close();
    };
    ClientDBConnection.prototype.scope = function (storeNames, mode, cb) {
        return __awaiter(this, void 0, void 0, function () {
            var prom, transaction, tx, retVal;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        prom = (0, createPromiseData_1.CreatePromiseData)();
                        transaction = this.db.transaction(storeNames, mode);
                        transaction.addEventListener("complete", function (ev) { prom.resolver(); });
                        transaction.addEventListener("error", function (ev) { prom.reject(ev.target.error); });
                        transaction.addEventListener("abort", function (ev) { prom.reject(ev.target.error); });
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, , 4, 5]);
                        tx = new clientDBTransactionDict[mode](transaction, storeNames);
                        return [4 /*yield*/, cb(tx)];
                    case 2:
                        retVal = _a.sent();
                        transaction.commit();
                        return [4 /*yield*/, prom.promise];
                    case 3:
                        _a.sent();
                        return [2 /*return*/, retVal];
                    case 4:
                        this.closeDB();
                        return [7 /*endfinally*/];
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    return ClientDBConnection;
}());
var ClientDB = /** @class */ (function () {
    function ClientDB(logger) {
        this.logger = logger;
    }
    ClientDB.prototype.listDatabases = function () {
        return __awaiter(this, void 0, void 0, function () {
            var dbList;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, indexedDB.databases()];
                    case 1:
                        dbList = _a.sent();
                        return [2 /*return*/, dbList.map(function (db) {
                                return {
                                    name: db.name,
                                    version: db.version,
                                };
                            })];
                }
            });
        });
    };
    ClientDB.prototype.openDB = function (dbName, version, options) {
        var _this = this;
        var prom = (0, createPromiseData_1.CreatePromiseData)();
        var request = indexedDB.open(dbName, version);
        request.addEventListener("error", function (ev) {
            var message = "error opening db: " + dbName + " v: " + version;
            var error = ev.target.error;
            _this.logger.log({ level: "error", message: message, error: error });
            prom.reject(error);
        });
        request.addEventListener("upgradeneeded", function (ev) { return __awaiter(_this, void 0, void 0, function () {
            var target, db, tx, ex_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.logger.log({ level: "warning", message: "db upgrading from: " + ev.oldVersion + " to " + ev.newVersion + "]" });
                        target = ev.target;
                        db = target.result;
                        tx = target.transaction;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, options.upgradeNeeded(new ClientDBUpgrader(db, tx), { new: ev.newVersion, old: ev.oldVersion })];
                    case 2:
                        _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        ex_1 = _a.sent();
                        this.logger.log({ level: "error", error: ex_1, message: "error on upgrade" });
                        tx.abort();
                        throw ex_1;
                    case 4: return [2 /*return*/];
                }
            });
        }); });
        request.addEventListener("success", function (ev) {
            var db = ev.target.result;
            prom.resolver(new ClientDBConnection(db));
        });
        return prom.promise;
    };
    ClientDB.prototype.deleteDB = function (dbName) {
        var _this = this;
        var prom = (0, createPromiseData_1.CreatePromiseData)();
        var request = indexedDB.deleteDatabase(dbName);
        request.addEventListener("error", function (ev) {
            var message = "error deleting db: " + dbName;
            var error = ev.target.error;
            _this.logger.log({ level: "error", message: message, error: error });
            prom.reject(error);
        });
        request.addEventListener("success", function (ev) {
            prom.resolver();
        });
    };
    return ClientDB;
}());
exports.ClientDB = ClientDB;
