// @stechquick/symphony-cli/lib/helpers/deploy/nanoDeployHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NanoDeployHelper = void 0;
const nanoPredeploy_1 = require("../../predeploy/src/nano/nanoPredeploy");
const nanoPostdeploy_1 = require("../../predeploy/src/nano/nanoPostdeploy");
const shellHelper_1 = require("../shellHelper");
const initHelper_1 = require("../initHelper");
const log_1 = require("@stechquick/symphony-common/lib/log");
const chalk_1 = require("../chalk");
class NanoDeployHelper {
    constructor(settings) {
        this.settings = settings;
    }
    async deploy() {
        if (!initHelper_1.InitHelper.AssertInFolder()) {
            return;
        }
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Deploy starting..."), timeFormat: "datetime" });
        let success = false;
        try {
            await nanoPredeploy_1.NanoPredeploy.execute(this.getReplaceWords());
            await this.nanoDeploy();
            success = true;
        }
        finally {
            nanoPostdeploy_1.NanoPostDeploy.execute();
            if (!success) {
                log_1.Logger.log({ message: chalk_1.Chalk.create("red.bold", "Deploy failed, ^^ search for errors up ^^"), timeFormat: "datetime" });
            }
            else {
                log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Deploy complete"), timeFormat: "datetime" });
            }
        }
    }
    getReplaceWords() {
        let data = { environmentalVars: "" };
        if (this.settings.nodeEnv) {
            data.environmentalVars = `ENV NODE_ENV=${this.settings.nodeEnv}`;
        }
        // //!!! Orbay: Bu böyle yönetilmemeli! Geçici yazıldı. Sanki app'in içinden bir yerden almalı bu bilgiyi???
        // switch (this.settings.environment) {
        //     case "openvpn":
        //         data.npm_registry = `# registry=http://nexus-service/repository/rally-npm/ | http://nexus.rally.softtech/repository/rally-npm/
        //         # email=npm_upload_user@softtech.com.tr
        //         # //nexus.rally.softtech/repository/rally-npm/:_authToken=NpmToken.8bcac3c0-8431-37d7-8c79-3b734756bea1
        //         # //nexus.rally.softtech/repository/npm-hosted/:_authToken=NpmToken.1201c29d-4df1-3acd-a2c6-0a839a902c52
        //         registry=http://nexus.rally.softtech/repository/rally-npm/
        //         email=npm_upload_user@softtech.com.tr
        //         //nexus.rally.softtech/repository/rally-npm/:_authToken=NpmToken.8bcac3c0-8431-37d7-8c79-3b734756bea1
        //         //nexus.rally.softtech/repository/npm-hosted/:_authToken=NpmToken.1201c29d-4df1-3acd-a2c6-0a839a902c52`;
        //         break;
        //     case "pulsevpn":
        //         data.npm_registry = `registry=http://nexus.softtech:8081/repository/softtech`;
        //         break;
        //     // case "remote":
        //     //     data.npm_registry = `# registry=http://nexus-service/repository/rally-npm/ | http://nexus.rally.softtech/repository/rally-npm/
        //     //     # email=npm_upload_user@softtech.com.tr
        //     //     # //nexus.rally.softtech/repository/rally-npm/:_authToken=NpmToken.8bcac3c0-8431-37d7-8c79-3b734756bea1
        //     //     # //nexus.rally.softtech/repository/npm-hosted/:_authToken=NpmToken.1201c29d-4df1-3acd-a2c6-0a839a902c52
        //     //     registry=http://nexus-service/repository/rally-npm/
        //     //     email=npm_upload_user@softtech.com.tr
        //     //     //nexus-service/repository/rally-npm/:_authToken=NpmToken.8bcac3c0-8431-37d7-8c79-3b734756bea1
        //     //     //nexus-service/repository/npm-hosted/:_authToken=NpmToken.1201c29d-4df1-3acd-a2c6-0a839a902c52`; 
        //     //     break;
        // }
        return data;
    }
    async fnContextDelete(contextName, fnFileName) {
        log_1.Logger.log({ message: "fnContext delete existing context...: " + contextName, timeFormat: "datetime" });
        await shellHelper_1.ShellHelper.ExecuteCommand(fnFileName, ["delete", "context", contextName]);
        log_1.Logger.log({ message: "fnContext deleted existing context: " + contextName, timeFormat: "datetime" });
    }
    async nanoDeploy() {
        // const packageJson = JSON.parse(fs.readFileSync('./package.json', 'utf8'));
        // //const ImageName="gcr.io/softtech-rally/symphony/isleasing:latest"
        // const appName = packageJson.name || "symphony-app";
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "docker build is starting"), timeFormat: "datetime" });
        await shellHelper_1.ShellHelper.ExecuteCommand('docker', ['build', '-t', this.settings.imageName, '.']);
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "docker build is finished"), timeFormat: "datetime" });
    }
}
exports.NanoDeployHelper = NanoDeployHelper;
