// @stechquick/algae/lib/quick/ITheme.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getDefaultFonts = exports.FontStyles = exports.FontWeights = void 0;
var FontWeights;
(function (FontWeights) {
    FontWeights[FontWeights["Thin - 100"] = 100] = "Thin - 100";
    FontWeights[FontWeights["Extra Light - 200"] = 200] = "Extra Light - 200";
    FontWeights[FontWeights["Light - 300"] = 300] = "Light - 300";
    FontWeights[FontWeights["Regular - 400"] = 400] = "Regular - 400";
    FontWeights[FontWeights["Medium - 500"] = 500] = "Medium - 500";
    FontWeights[FontWeights["SemiBold - 600"] = 600] = "SemiBold - 600";
    FontWeights[FontWeights["Bold - 700"] = 700] = "Bold - 700";
    FontWeights[FontWeights["Extra Bold - 800"] = 800] = "Extra Bold - 800";
    FontWeights[FontWeights["Black - 900"] = 900] = "Black - 900";
    FontWeights[FontWeights["Weight - 1000"] = 1000] = "Weight - 1000";
})(FontWeights = exports.FontWeights || (exports.FontWeights = {}));
var FontStyles;
(function (FontStyles) {
    FontStyles[FontStyles["Italic"] = 0] = "Italic";
})(FontStyles = exports.FontStyles || (exports.FontStyles = {}));
var getDefaultFonts = function () {
    var fonts = [
        // {
        //   fontFamily: "Arial",
        //   italic: true,
        //   fontWeights: [FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Extra Bold - 800"]]]
        // },
        {
            fontFamily: "Inter",
            fontWeights: [FontWeights[FontWeights["Thin - 100"]], FontWeights[FontWeights["Extra Light - 200"]], FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Extra Bold - 800"]], FontWeights[FontWeights["Black - 900"]]]
        },
        {
            fontFamily: "Lato",
            fontWeights: [FontWeights[FontWeights["Thin - 100"]], FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Black - 900"]]],
            italic: true
        },
        {
            fontFamily: "Montserrat",
            fontWeights: [FontWeights[FontWeights["Thin - 100"]], FontWeights[FontWeights["Extra Light - 200"]], FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Extra Bold - 800"]], FontWeights[FontWeights["Black - 900"]]],
            italic: true
        },
        {
            fontFamily: "Nunito",
            fontWeights: [FontWeights[FontWeights["Extra Light - 200"]], FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Extra Bold - 800"]], FontWeights[FontWeights["Black - 900"]], FontWeights[FontWeights["Weight - 1000"]]],
            italic: true
        },
        {
            fontFamily: "Open Sans",
            fontWeights: [FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Extra Bold - 800"]]],
            italic: true
        },
        {
            fontFamily: "Oswald",
            fontWeights: [FontWeights[FontWeights["Extra Light - 200"]], FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Bold - 700"]]]
        }, {
            fontFamily: "Poppins",
            italic: true,
            fontWeights: [FontWeights[FontWeights["Thin - 100"]], FontWeights[FontWeights["Extra Light - 200"]], FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Extra Bold - 800"]], FontWeights[FontWeights["Black - 900"]]]
        },
        {
            fontFamily: "Raleway",
            italic: true,
            fontWeights: [FontWeights[FontWeights["Thin - 100"]], FontWeights[FontWeights["Extra Light - 200"]], FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Extra Bold - 800"]], FontWeights[FontWeights["Black - 900"]]]
        },
        {
            fontFamily: "Roboto",
            italic: true,
            fontWeights: [FontWeights[FontWeights["Thin - 100"]], FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Black - 900"]]]
        },
        {
            fontFamily: "Rubik",
            italic: true,
            fontWeights: [FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Extra Bold - 800"]], FontWeights[FontWeights["Black - 900"]]]
        },
        // {
        //   fontFamily: "Segoe UI",
        //   fontWeights: [FontWeights[FontWeights["Thin - 100"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["Medium - 500"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Extra Bold - 800"]]],
        //   italic: true
        // },
        {
            fontFamily: "Source Sans Pro",
            fontWeights: [FontWeights[FontWeights["Extra Light - 200"]], FontWeights[FontWeights["Light - 300"]], FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["SemiBold - 600"]], FontWeights[FontWeights["Bold - 700"]], FontWeights[FontWeights["Black - 900"]]],
            italic: true
        },
        // {
        //   fontFamily: "Verdana",
        //   fontWeights: [FontWeights[FontWeights["Regular - 400"]], FontWeights[FontWeights["SemiBold - 600"]]],
        //   italic: true
        // }
    ];
    var defaultFonts = fonts.flatMap(function (font) {
        if (font.italic) {
            return [font, { fontFamily: font.fontFamily + " Italic", fontWeights: font.fontWeights }];
        }
        return font;
    });
    return defaultFonts;
};
exports.getDefaultFonts = getDefaultFonts;
