// @stechquick/algae/lib/quick/quickSettingsBundler.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.QuickSettingsBundler = void 0;
var fs = __importStar(require("fs"));
var path_1 = __importDefault(require("path"));
var fileHelper_1 = require("../helpers/fileHelper");
var integrationsHelper_1 = require("../helpers/integrationsHelper");
var objectHelper_1 = require("../helpers/objectHelper");
var qJsonParser_1 = require("../helpers/qJsonParser");
var stringHelper_1 = require("../helpers/stringHelper");
var QAnimationTemplate_1 = require("./QAnimationTemplate");
var fsPromises = fs.promises;
var QuickSettingsBundler = /** @class */ (function () {
    function QuickSettingsBundler() {
    }
    QuickSettingsBundler.prototype.bundleSettings = function (distFolder, bundleContainer) {
        var _a, _b, _c;
        return __awaiter(this, void 0, void 0, function () {
            var settingsPath, localProxyPath, assetListPath, settingsContent, localProxyContent, assetListContent, _d, settings, localProxy, assetList, lottieFilePlateauUIObject, combinedSettings, settingsWeb, fileBody, concattedYamlStringWeb, settingsIAM, concattedYamlStringIAM, mobileCss, concatenatedValues, style, _i, _e, themeKey, theme, _f, _g, cssFileKey, cssFile, resultString, settingsMobile, iamV2PlateauObject, assetListPlateauObject, concattedYamlStringMobile, infoFile, concattedYamlStringInfo;
            return __generator(this, function (_h) {
                switch (_h.label) {
                    case 0:
                        settingsPath = path_1.default.join(distFolder, "/client", "/settings/settings.yaml");
                        localProxyPath = path_1.default.join(distFolder, "/client", "/settings/localProxy.yaml");
                        assetListPath = path_1.default.join(distFolder, "/client", "/static/css/assetList.yaml");
                        return [4 /*yield*/, fsPromises.readFile(settingsPath, { encoding: "utf-8" })];
                    case 1:
                        settingsContent = _h.sent();
                        return [4 /*yield*/, (0, fileHelper_1.tryGetContents)(localProxyPath)];
                    case 2:
                        localProxyContent = _h.sent();
                        if (!fs.existsSync(assetListPath)) return [3 /*break*/, 4];
                        return [4 /*yield*/, fsPromises.readFile(assetListPath, { encoding: "utf-8" })];
                    case 3:
                        _d = _h.sent();
                        return [3 /*break*/, 5];
                    case 4:
                        _d = undefined;
                        _h.label = 5;
                    case 5:
                        assetListContent = _d;
                        settings = bundleContainer.yaml.load(settingsContent);
                        localProxy = localProxyContent ? bundleContainer.yaml.load(localProxyContent) : undefined;
                        assetList = assetListContent ? bundleContainer.yaml.load(assetListContent) : undefined;
                        return [4 /*yield*/, this.checkandsetLottieFile(settings)];
                    case 6:
                        lottieFilePlateauUIObject = _h.sent();
                        if (settings.appSettings && settings.appSettings.integrationsSettings) {
                            integrationsHelper_1.IntegrationsHelper.arrangeIndexHtml(settings.appSettings.integrationsSettings, path_1.default.join(distFolder, "/client/index.html"));
                        }
                        // CopyFolder("./static", "./dist/client/static", undefined, 
                        // { distinctFileExtensions: [".qjson"], distinctFilesCb: this.minifyQjson, context: this }); // copies static folder. if encountered any file with ".qjson" extension, calls minifyQjson
                        return [4 /*yield*/, this.replaceTsFilesAndSettings(distFolder, {
                                "componentListPath": "/static/js/componentList.ts",
                                "containerServicesPath": "/static/js/containerServices.ts",
                            }, settings, bundleContainer)];
                    case 7:
                        // CopyFolder("./static", "./dist/client/static", undefined, 
                        // { distinctFileExtensions: [".qjson"], distinctFilesCb: this.minifyQjson, context: this }); // copies static folder. if encountered any file with ".qjson" extension, calls minifyQjson
                        _h.sent();
                        return [4 /*yield*/, this.combineSettings(distFolder, {
                                settings: settings,
                                localProxy: localProxy,
                            }, {
                                componentList: { type: "object", settingsYamlKey: "componentListPath" },
                                containerServices: { type: "js", settingsYamlKey: "containerServicesPath" },
                                pipeline: { format: "qjson", type: "js", settingsYamlKey: "pipelineqjson" },
                                alert: { format: "qjson", type: "js", settingsYamlKey: "alertqjson" },
                                loading: { format: "qjson", type: "js", settingsYamlKey: "loadingqjson" },
                                loadingLottie: { format: "lottie", type: "js", settingsYamlKey: "loadingfile" },
                                globalLocalization: { format: "qjson", type: "js", settingsYamlKey: "globalLocalizationqjson" },
                                rootqjson: { format: "qjson", type: "js", settingsYamlKey: "rootqjson" },
                                style: { type: "style", settingsYamlKey: "cssPath" },
                                font: { type: "font", settingsYamlKey: "fontPath" },
                                css: { type: "style", settingsYamlKey: "css" },
                                container: { type: "js", settingsYamlKey: "container" },
                            }, bundleContainer)];
                    case 8:
                        combinedSettings = _h.sent();
                        settingsWeb = {
                            settings: combinedSettings.settings,
                            componentList: combinedSettings.componentList,
                            containerServices: combinedSettings.containerServices,
                            pipeline: combinedSettings.pipeline,
                            alert: combinedSettings.alert,
                            loading: lottieFilePlateauUIObject || combinedSettings.loading,
                            globalLocalization: combinedSettings.globalLocalization,
                            rootqjson: combinedSettings.rootqjson,
                            localProxy: combinedSettings.localProxy,
                            style: combinedSettings.style,
                            container: combinedSettings.container,
                        };
                        fileBody = "var assetObj = "
                            + JSON.stringify(assetList, null, '\t')
                            + "\nif (typeof exports === 'object' && typeof module === 'object'){\n \tmodule.exports = assetObj;\n}else{ \n\twindow['plateauUIAsset'] = assetObj;\n}";
                        return [4 /*yield*/, fsPromises.writeFile(path_1.default.join(distFolder, "./client", "/static/css/assetList.js"), fileBody, { encoding: "utf-8" })];
                    case 9:
                        _h.sent();
                        concattedYamlStringWeb = "var settings = "
                            + JSON.stringify(settingsWeb, null, '\t')
                            + "\nif (typeof exports === 'object' && typeof module === 'object'){\n \tmodule.exports = settings;\n}else{ \n\twindow['plateauUISettings'] = settings;\n}";
                        return [4 /*yield*/, fsPromises.writeFile(path_1.default.join(distFolder, "./client", "/settings/settings_web.js"), concattedYamlStringWeb, { encoding: "utf-8" })];
                    case 10:
                        _h.sent();
                        settingsIAM = { iam: (_a = settings.backendInf) === null || _a === void 0 ? void 0 : _a.iam, iamV2: settings.iamV2 };
                        concattedYamlStringIAM = "const plateauUIIAMOptions = "
                            + JSON.stringify(settingsIAM, null, '\t')
                            + "\nif (typeof exports === 'object' && typeof module === 'object'){\n \tmodule.exports = plateauUIIAMOptions;\n}else{ \n\twindow['plateauUIIAMOptions'] = plateauUIIAMOptions;\n}";
                        return [4 /*yield*/, fsPromises.writeFile(path_1.default.join(distFolder, "./client", "/settings/settings_iam.js"), concattedYamlStringIAM, { encoding: "utf-8" })];
                    case 11:
                        _h.sent();
                        mobileCss = { value: "", type: "style" };
                        concatenatedValues = new Set();
                        style = (_b = combinedSettings.style) === null || _b === void 0 ? void 0 : _b.value;
                        for (_i = 0, _e = Object.keys(style); _i < _e.length; _i++) {
                            themeKey = _e[_i];
                            theme = style[themeKey];
                            if (themeKey.includes("css")) {
                                concatenatedValues.add(theme.value + " \r\n");
                            }
                            else {
                                for (_f = 0, _g = Object.keys(theme); _f < _g.length; _f++) {
                                    cssFileKey = _g[_f];
                                    cssFile = theme[cssFileKey];
                                    concatenatedValues.add(cssFile.value + " \r\n");
                                }
                            }
                        }
                        resultString = Array.from(concatenatedValues).join(" ");
                        mobileCss.value = resultString;
                        settingsMobile = {
                            settings: combinedSettings.settings,
                            pipeline: combinedSettings.pipeline,
                            alert: combinedSettings.alert,
                            loading: (_c = combinedSettings.loadingLottie) !== null && _c !== void 0 ? _c : combinedSettings.loading,
                            globalLocalization: combinedSettings.globalLocalization,
                            rootqjson: combinedSettings.rootqjson,
                            css: mobileCss,
                            container: combinedSettings.container,
                        };
                        iamV2PlateauObject = { type: "object", value: settings.iamV2 };
                        assetListPlateauObject = { type: "object", value: assetList };
                        concattedYamlStringMobile = JSON.stringify(__assign(__assign({}, settingsMobile), { iamV2: iamV2PlateauObject, assetList: assetListPlateauObject }), null, '\t');
                        return [4 /*yield*/, fsPromises.writeFile(path_1.default.join(distFolder, "./client", "/settings/settings_mobile.json"), concattedYamlStringMobile, { encoding: "utf-8" })];
                    case 12:
                        _h.sent();
                        infoFile = {
                            buildTime: new Date().toISOString(),
                        };
                        concattedYamlStringInfo = JSON.stringify(infoFile, null, '\t');
                        return [4 /*yield*/, fsPromises.writeFile(path_1.default.join(distFolder, "./client", "/info.json"), concattedYamlStringInfo, { encoding: "utf-8" })];
                    case 13:
                        _h.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    QuickSettingsBundler.prototype.replaceTsFilesAndSettings = function (distFolder, paths, settings, container) {
        return __awaiter(this, void 0, void 0, function () {
            var obj, proms;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        obj = {
                            componentListPath: function (fileData) { return __awaiter(_this, void 0, void 0, function () {
                                var retVal, fileBody;
                                return __generator(this, function (_a) {
                                    retVal = { fileName: fileData.fileName.replace(".ts", ".js"), fileBody: fileData.fileBody };
                                    fileBody = stringHelper_1.StringHelper.substringBetween(fileData.fileBody, "//#region JS Section");
                                    retVal.fileBody = retVal.fileBody.replace('window["QEditorComponentOptions"]', 'window["plateauUIComponentOptions"]'); // for old versions of qui 
                                    if (!fileBody) {
                                        return [2 /*return*/, retVal];
                                    }
                                    retVal.fileBody = fileBody.replace(": ComponentOptions =", " =");
                                    return [2 /*return*/, retVal];
                                });
                            }); },
                            containerServicesPath: function (fileData) { return __awaiter(_this, void 0, void 0, function () {
                                var retVal, fileBody;
                                return __generator(this, function (_a) {
                                    retVal = { fileName: fileData.fileName.replace(".ts", ".js"), fileBody: fileData.fileBody };
                                    fileBody = stringHelper_1.StringHelper.substringBetween(fileData.fileBody, "//#region JS Section");
                                    if (!fileBody) {
                                        return [2 /*return*/, retVal];
                                    }
                                    retVal.fileBody = fileBody;
                                    return [2 /*return*/, retVal];
                                });
                            }); },
                        };
                        proms = Object.keys(paths).map(function (pathKey) { return __awaiter(_this, void 0, void 0, function () {
                            var fileName, oldFileName, fileBody, newFileData;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        fileName = settings[pathKey] || paths[pathKey];
                                        oldFileName = path_1.default.join(distFolder, "/client", fileName);
                                        return [4 /*yield*/, (0, fileHelper_1.tryGetContents)(oldFileName)];
                                    case 1:
                                        fileBody = _a.sent();
                                        if (!fileBody) {
                                            return [2 /*return*/];
                                        }
                                        return [4 /*yield*/, obj[pathKey]({ fileBody: fileBody, fileName: fileName })];
                                    case 2:
                                        newFileData = _a.sent();
                                        return [4 /*yield*/, fsPromises.writeFile(path_1.default.join(distFolder, "/client", newFileData.fileName), newFileData.fileBody, { encoding: "utf-8" })];
                                    case 3:
                                        _a.sent();
                                        return [4 /*yield*/, fsPromises.unlink(oldFileName)];
                                    case 4:
                                        _a.sent();
                                        settings[pathKey] = newFileData.fileName;
                                        return [2 /*return*/];
                                }
                            });
                        }); });
                        return [4 /*yield*/, Promise.all(proms)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        });
    };
    QuickSettingsBundler.prototype.tryGetCssFontContent = function (distFolder, itemValueInSettingsYaml, result, contentType) {
        return __awaiter(this, void 0, void 0, function () {
            var _i, itemValueInSettingsYaml_1, item, _a, _b, key, basePath, itemContent, subString, fileName;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        result = result || {};
                        if (!itemValueInSettingsYaml) return [3 /*break*/, 12];
                        if (!(typeof itemValueInSettingsYaml != "string")) return [3 /*break*/, 10];
                        if (!Array.isArray(itemValueInSettingsYaml)) return [3 /*break*/, 5];
                        _i = 0, itemValueInSettingsYaml_1 = itemValueInSettingsYaml;
                        _c.label = 1;
                    case 1:
                        if (!(_i < itemValueInSettingsYaml_1.length)) return [3 /*break*/, 4];
                        item = itemValueInSettingsYaml_1[_i];
                        return [4 /*yield*/, this.tryGetCssFontContent(distFolder, item, result, contentType)];
                    case 2:
                        _c.sent();
                        _c.label = 3;
                    case 3:
                        _i++;
                        return [3 /*break*/, 1];
                    case 4: return [3 /*break*/, 9];
                    case 5:
                        if (!(Object.keys(itemValueInSettingsYaml).length > 0)) return [3 /*break*/, 9];
                        _a = 0, _b = Object.keys(itemValueInSettingsYaml);
                        _c.label = 6;
                    case 6:
                        if (!(_a < _b.length)) return [3 /*break*/, 9];
                        key = _b[_a];
                        return [4 /*yield*/, this.tryGetCssFontContent(distFolder, itemValueInSettingsYaml[key], result[key] = {}, contentType)];
                    case 7:
                        _c.sent();
                        _c.label = 8;
                    case 8:
                        _a++;
                        return [3 /*break*/, 6];
                    case 9: return [3 /*break*/, 12];
                    case 10:
                        basePath = path_1.default.join(distFolder, "/client", itemValueInSettingsYaml);
                        return [4 /*yield*/, (0, fileHelper_1.tryGetContents)(basePath)];
                    case 11:
                        itemContent = _c.sent();
                        subString = basePath.includes("/") ? "/" : "\\";
                        fileName = basePath.substring(basePath.lastIndexOf(subString) + 1);
                        result[fileName] = itemContent ? { value: itemContent, type: "object" } : { value: itemValueInSettingsYaml, type: "path" };
                        _c.label = 12;
                    case 12: return [2 /*return*/];
                }
            });
        });
    };
    QuickSettingsBundler.prototype.combineSettings = function (distFolder, items, plateauUIObjectConfigs, container) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            var settingsTotal, proms, cssDict, fontDict;
            var _this = this;
            return __generator(this, function (_c) {
                switch (_c.label) {
                    case 0:
                        settingsTotal = {
                            settings: { value: items.settings, type: "object" },
                            localProxy: items.localProxy ? { value: items.localProxy, type: "object" } : undefined
                        };
                        proms = (0, objectHelper_1.objectKeys)(plateauUIObjectConfigs).filter(function (key) { return plateauUIObjectConfigs[key].type != "style" && plateauUIObjectConfigs[key].type != "font"; }).map(function (key) { return __awaiter(_this, void 0, void 0, function () {
                            var itemConfig, itemPath, itemContent, objRef, module, exports;
                            var _a;
                            return __generator(this, function (_b) {
                                switch (_b.label) {
                                    case 0:
                                        itemConfig = plateauUIObjectConfigs[key];
                                        itemPath = (_a = items.settings[itemConfig.settingsYamlKey]) !== null && _a !== void 0 ? _a : items.settings[itemConfig.settingsYamlKey.toLocaleLowerCase("tr")];
                                        if (!itemPath) {
                                            console.log(key + " settings does not exist, skipping");
                                            return [2 /*return*/];
                                        }
                                        console.log("checking " + key + " file");
                                        return [4 /*yield*/, (0, fileHelper_1.tryGetContents)(path_1.default.join(distFolder, "/client", itemPath))];
                                    case 1:
                                        itemContent = _b.sent();
                                        if (!itemContent) {
                                            console.log(key + " file does not exist, skipping");
                                            settingsTotal[key] = { type: "path", value: itemPath, format: itemConfig.format };
                                            return [2 /*return*/];
                                        }
                                        // if (key == "lottieanimation"){
                                        //     const lottieJsonObj = JSON.parse(itemContent);
                                        //     for (let item of lottieJsonObj.pJSON){
                                        //         item.P.source.H = lottieContent
                                        //     }
                                        //     for (let item of lottieJsonObj.cJSON){
                                        //         item.P.source.H = lottieContent
                                        //     }
                                        //     itemContent = JSON.stringify(lottieJsonObj,null,2);
                                        //     return;
                                        // }
                                        console.log(key + " file exist");
                                        switch (itemConfig.type) {
                                            default: throw new Error("unexpected type: " + itemConfig.type);
                                            case "js":
                                                objRef = settingsTotal[key] = { type: itemConfig.type, value: itemContent };
                                                if (itemConfig.format == "qjson") {
                                                    objRef.value = new qJsonParser_1.QJsonParser().minifyQjson(objRef.value, items.settings.keepSourceMappingURL);
                                                    objRef.format = "qjson";
                                                }
                                                if (itemConfig.format) {
                                                    objRef.format = itemConfig.format;
                                                }
                                                break;
                                            case "object":
                                                module = {};
                                                exports = {};
                                                (new Function("module", "exports", itemContent)(module, exports));
                                                settingsTotal[key] = { value: module === null || module === void 0 ? void 0 : module.exports, type: "object" };
                                                break;
                                        }
                                        return [2 /*return*/];
                                }
                            });
                        }); });
                        return [4 /*yield*/, Promise.all(proms)];
                    case 1:
                        _c.sent();
                        cssDict = {};
                        return [4 /*yield*/, this.tryGetCssFontContent(distFolder, (_a = items.settings[plateauUIObjectConfigs.style.settingsYamlKey]) !== null && _a !== void 0 ? _a : items.settings[plateauUIObjectConfigs.style.settingsYamlKey.toLocaleLowerCase("tr")], cssDict, 'style')];
                    case 2:
                        _c.sent();
                        settingsTotal["style"] = { value: cssDict, type: "object" };
                        fontDict = {};
                        return [4 /*yield*/, this.tryGetCssFontContent(distFolder, (_b = items.settings[plateauUIObjectConfigs.font.settingsYamlKey]) !== null && _b !== void 0 ? _b : items.settings[plateauUIObjectConfigs.font.settingsYamlKey.toLocaleLowerCase("tr")], fontDict, 'font')];
                    case 3:
                        _c.sent();
                        settingsTotal["font"] = { value: fontDict, type: "object" };
                        return [2 /*return*/, settingsTotal];
                }
            });
        });
    };
    QuickSettingsBundler.prototype.checkandsetLottieFile = function (settings) {
        return __awaiter(this, void 0, void 0, function () {
            var QAnimationQjsonFolder, QAnimationQjsonFile, lottiePath, loadingFileContent;
            return __generator(this, function (_a) {
                QAnimationQjsonFolder = "./settings/qjsons";
                QAnimationQjsonFile = "/QAnimation.qjson";
                if (!settings.loadingfile) {
                    console.log("Lottie File is not provided!");
                    return [2 /*return*/];
                }
                lottiePath = settings.loadingfile;
                loadingFileContent = QAnimationTemplate_1.QAnimationTemplate.qanimationGenerator(lottiePath);
                settings.loadingqjson = path_1.default.join(QAnimationQjsonFolder, QAnimationQjsonFile); //TODO: gerekyok silinebilir.
                return [2 /*return*/, { format: "qjson", type: "js", value: loadingFileContent }];
            });
        });
    };
    return QuickSettingsBundler;
}());
exports.QuickSettingsBundler = QuickSettingsBundler;
;
;
