// @stechquick/symphony-cli/lib/commands/mui-create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const microUIHelper_1 = require("../helpers/microUIHelper");
const npmHelper_1 = require("../helpers/npmHelper");
const log_1 = require("@stechquick/symphony-common/lib/log");
const chalk_1 = require("../helpers/chalk");
class CreateMicroUI extends command_1.Command {
    async run() {
        const { flags } = this.parse(CreateMicroUI);
        const { args } = this.parse(CreateMicroUI);
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Creating micro ui, please wait..."), timeFormat: "datetime" });
        await new microUIHelper_1.MicroUIHelper().CreateMicroUI(args["name"], `${flags.desc}`);
        npmHelper_1.NpmHelper.install({ folderName: args["name"] + "-ui" });
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Micro ui created successfully.") });
    }
}
exports.default = CreateMicroUI;
CreateMicroUI.description = 'MicroUI command to create a new Micro UI project';
CreateMicroUI.examples = [
    `symphony-cli mui-create myUI -d="My Micro UI project"`
];
CreateMicroUI.args = [
    {
        name: 'name',
        required: true,
        description: 'input microui name',
    }
];
CreateMicroUI.flags = {
    desc: command_1.flags.string({
        char: 'd',
        description: 'description to project',
        hidden: false,
        multiple: false,
        default: '',
        required: false,
    }),
};
