// @stechquick/symphony-cli/lib/commands/nc-create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const ncHelper_1 = require("../helpers/namedComponent/ncHelper");
const npmHelper_1 = require("../helpers/npmHelper");
const chalk_1 = require("../helpers/chalk");
const log_1 = require("@stechquick/symphony-common/lib/log");
class CreateNC extends command_1.Command {
    async run() {
        const { flags, args } = this.parse(CreateNC);
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Creating named component template. Please wait..."), timeFormat: "datetime" });
        await new ncHelper_1.NCHelper().CreateNC(args["name"], `${flags.desc}`);
        npmHelper_1.NpmHelper.install({ folderName: args["name"] });
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Named component template created successfully."), timeFormat: "datetime" });
    }
}
exports.default = CreateNC;
CreateNC.description = 'Command for create a named component template project.';
CreateNC.examples = [
    `symphony-cli nc-create nc_example -d="my named component example"`
];
CreateNC.args = [
    {
        name: 'name',
        required: true,
        description: 'input project name',
    }
];
CreateNC.flags = {
    desc: command_1.flags.string({
        char: 'd',
        description: 'description to named component',
        hidden: false,
        multiple: false,
        default: '',
        required: false,
    }),
};
