// @stechquick/symphony-cli/lib/commands/qui-create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const createuiHelper_1 = require("../helpers/ui/createuiHelper");
const npmHelper_1 = require("../helpers/npmHelper");
const log_1 = require("@stechquick/symphony-common/lib/log");
const chalk_1 = require("../helpers/chalk");
class CreateUI extends command_1.Command {
    async run() {
        const { flags } = this.parse(CreateUI);
        const { args } = this.parse(CreateUI);
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Creating UI. Please wait..."), timeFormat: "datetime" });
        let createResponse = await new createuiHelper_1.CreateUIHelper().CreateUI(args["name"], `${flags.desc}`, `${flags.token}`, `${flags.registry}`);
        if (createResponse == false) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "UI creation failed due to an error. Please review the error message for more details."), timeFormat: "datetime" });
        }
        else {
            npmHelper_1.NpmHelper.install({ folderName: args["name"] });
            log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "UI created successfully."), timeFormat: "datetime" });
        }
    }
}
exports.default = CreateUI;
CreateUI.description = 'Plateau-UI command to create a new Plateau UI project';
CreateUI.examples = [
    `symphony-cli qui-create myUI -t=TOKEN`,
    `symphony-cli qui-create myUI -t=TOKEN -d="My Plateau UI project"`,
    `symphony-cli qui-create myUI -r=REGISTERY`,
    `symphony-cli qui-create myUI -r=REGISTERY -d="My Plateau UI project"`
];
CreateUI.args = [
    {
        name: 'name',
        required: true,
        description: 'input Plateau UI name',
    }
];
CreateUI.flags = {
    desc: command_1.flags.string({
        char: 'd',
        description: 'description to project',
        hidden: false,
        multiple: false,
        default: '',
        required: false,
    }),
    token: command_1.flags.string({
        char: "t",
        description: "npm token for @stechquick",
        hidden: false,
        multiple: false,
        default: '',
        required: false,
    }),
    registry: command_1.flags.string({
        char: "r",
        description: "your registery address (example: nexus address)",
        hidden: false,
        multiple: false,
        default: '',
        required: false,
    })
};
