// @stechquick/symphony-cli/lib/commands/sym-create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const initHelper_1 = require("../helpers/initHelper");
const npmHelper_1 = require("../helpers/npmHelper");
const chalk_1 = require("../helpers/chalk");
const log_1 = require("@stechquick/symphony-common/lib/log");
class CreateApp extends command_1.Command {
    async run() {
        const { flags } = this.parse(CreateApp);
        const { args } = this.parse(CreateApp);
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Application " + args["name"] + " creating. Please wait...") });
        let createResponse = await new initHelper_1.InitHelper().AddApp(args["name"], `${flags.desc}`, `${flags.token}`, `${flags.registry}`);
        if (createResponse == false) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Application " + args["name"] + " creation failed due to an error. Please review the error message for more details."), timeFormat: "datetime" });
        }
        else {
            npmHelper_1.NpmHelper.install({ folderName: args["name"] });
            log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Application " + args["name"] + " created successfully."), timeFormat: "datetime" });
        }
    }
}
exports.default = CreateApp;
CreateApp.description = 'Plateau-Symphony command to create a Symphony App';
CreateApp.examples = [
    `symphony-cli sym-create havale -t=TOKEN`,
    `symphony-cli sym-create havale -t=TOKEN -d="Havale işlemi"`,
    `symphony-cli sym-create havale -r=REGISTERY`,
    `symphony-cli sym-create havale -r=REGISTERY -d="Havale işlemi"`
];
CreateApp.args = [
    {
        name: 'name',
        required: true,
        description: 'input application name',
    }
];
CreateApp.flags = {
    desc: command_1.flags.string({
        char: 'd',
        description: 'description to application',
        hidden: false,
        multiple: false,
        default: '',
        required: false,
    }),
    token: command_1.flags.string({
        char: "t",
        description: "npm token for @stechquick",
        hidden: false,
        multiple: false,
        default: '',
        required: false,
    }),
    registry: command_1.flags.string({
        char: "r",
        description: "your registery address (example: nexus address)",
        hidden: false,
        multiple: false,
        default: '',
        required: false,
    })
};
