// @stechquick/symphony-cli/lib/commands/sym-fn-deploy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const fnDeployHelper_1 = require("../helpers/deploy/fnDeployHelper");
const log_1 = require("@stechquick/symphony-common/lib/log");
const chalk_1 = require("../helpers/chalk");
class FnDeploy extends command_1.Command {
    async run() {
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Checking deployment parameters, please wait...") });
        const deploySettings = this.getSettings();
        if (!deploySettings) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("red.bold", "Deploy stopped.") });
            process.exit(12);
        }
        this.appendEnvIndependentSettings(deploySettings);
        await new fnDeployHelper_1.FnDeployHelper(deploySettings).deploy();
    }
    getSettings() {
        const { args } = this.parse(FnDeploy);
        switch (args.environment) {
            case "remote":
                return this.getRemoteSettings();
            case "play":
                return FnDeploy.defaultSettings;
            case "local":
                return { environment: "local" };
        }
    }
    appendEnvIndependentSettings(deploySettings) {
        const { flags } = this.parse(FnDeploy);
        const nodeEnv = flags.node_env || process.env["NODE_ENV"] || process.env["node_env"];
        if (!nodeEnv) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("yellow", "Warning: There is no 'environment name'. Defaulting to: default.") });
        }
        deploySettings.nodeEnv = nodeEnv;
    }
    getRemoteSettings() {
        const { flags } = this.parse(FnDeploy);
        const dockerUrl = flags.d_url || process.env["d_url"];
        if (!dockerUrl) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("yellow", "Warning: There is no 'docker url' definition!") });
        }
        const dockerUser = flags.d_user || process.env["d_user"];
        if (!dockerUser) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("yellow", "Warning: There is no 'docker username' definition!") });
        }
        const dockerPass = flags.d_pass || process.env["d_pass"];
        if (!dockerPass) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("yellow", "Warning: There is no 'docker pasword' definition!") });
        }
        const fnUrl = flags.f_url || process.env["f_url"];
        if (!fnUrl) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("yellow", "Warning: There is no 'fn api url' definition!") });
        }
        if (!(dockerUrl && dockerUser && dockerPass && fnUrl)) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("yellow", "Plase check your deployment parameters!") });
            return;
        }
        return {
            environment: "remote",
            dockerSettings: {
                url: dockerUrl,
                username: dockerUser,
                password: dockerPass
            },
            fnSettings: {
                apiUrl: fnUrl
            }
        };
    }
}
exports.default = FnDeploy;
FnDeploy.description = 'Plateau-Symphony fn command to deploy the all functions in your app';
FnDeploy.examples = [
    'symphony-cli sym-fn-deploy local',
    'symphony-cli sym-fn-deploy play',
    'symphony-cli sym-fn-deploy remote',
    'symphony-cli sym-fn-deploy remote -d MyDockerUrl -u MyDockerUser -p MyDockerPassword -f MyFnApiUrl',
    'symphony-cli sym-fn-deploy remote --d_url MyDockerUrl --d_user MyDockerUser --d_pass MyDockerPassword --f_url MyFnApiUrl',
    'symphony-cli sym-fn-deploy remote --d_url="MyDockerUrl" --d_user="MyDockerUser" --d_pass="MyDockerPassword" --f_url="MyFnApiUrl"'
    // Chalk.create("green", 'symphony-cli sym-deploy local'),
    // Chalk.create("red", 'symphony-cli sym-deploy play'),
    // Chalk.create("yellow", 'symphony-cli sym-deploy remote'),
    // Chalk.create("yellow", 'symphony-cli sym-deploy remote -d MyDockerUrl -u MyDockerUser -p MyDockerPassword -f MyFnApiUrl'),
    // Chalk.create("yellow", 'symphony-cli sym-deploy remote --d_url MyDockerUrl --d_user MyDockerUser --d_pass MyDockerPassword --f_url MyFnApiUrl'),
    // Chalk.create("yellow", 'symphony-cli sym-deploy remote --d_url="MyDockerUrl" --d_user="MyDockerUser" --d_pass="MyDockerPassword" --f_url="MyFnApiUrl"')
];
FnDeploy.args = [
    {
        name: 'environment',
        required: true,
        description: `${chalk_1.Chalk.create("bold", 'deploy environment: ')} 
                ${chalk_1.Chalk.create("green.bold", '[local: your local]')} 
                ${chalk_1.Chalk.create("red.bold", '[play: playground environment]')} 
                ${chalk_1.Chalk.create("yellow.bold", '[remote: remote server]')}`,
        hidden: false,
        options: ["local", "play", "remote"]
    }
];
FnDeploy.flags = {
    d_url: command_1.flags.string({
        description: 'docker url to deploy',
        hidden: false,
        multiple: false,
        default: '',
        required: false,
        char: 'd',
    }),
    d_user: command_1.flags.string({
        description: 'docker username to deploy',
        hidden: false,
        multiple: false,
        default: '',
        required: false,
        char: 'u',
    }),
    d_pass: command_1.flags.string({
        description: 'docker password for this username to deploy',
        hidden: false,
        multiple: false,
        default: '',
        required: false,
        char: 'p',
    }),
    f_url: command_1.flags.string({
        description: 'fn api url to deploy',
        hidden: false,
        multiple: false,
        default: '',
        required: false,
        char: 'f',
    }),
    node_env: command_1.flags.string({
        description: 'symphony environment name',
        hidden: false,
        multiple: false,
        default: undefined,
        required: false,
        char: 'e',
    }),
};
FnDeploy.defaultSettings = {
    environment: "play",
    dockerSettings: {
        url: "",
        username: "symphonydemo",
        password: "f40ddfd4-aad6-4a6a-b3a9-fb54847fb60a"
    },
    fnSettings: {
        apiUrl: "http://10.223.2.98:80"
    }
};
