// @stechquick/symphony-cli/lib/commands/sym-nano-deploy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const nanoDeployHelper_1 = require("../helpers/deploy/nanoDeployHelper");
const log_1 = require("@stechquick/symphony-common/lib/log");
const chalk_1 = require("../helpers/chalk");
class NanoDeploy extends command_1.Command {
    async run() {
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Checking deployment parameters, please wait...") });
        const { flags } = this.parse(NanoDeploy);
        const deploySettings = { imageName: flags.imagename };
        if (!deploySettings) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("red.bold", "Deploy stopped.") });
            process.exit(12);
        }
        this.appendEnvIndependentSettings(deploySettings);
        await new nanoDeployHelper_1.NanoDeployHelper(deploySettings).deploy();
    }
    appendEnvIndependentSettings(deploySettings) {
        const { flags } = this.parse(NanoDeploy);
        const nodeEnv = flags.node_env || process.env["NODE_ENV"] || process.env["node_env"];
        if (!nodeEnv) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("yellow", "Warning: There is no 'environment name'. Defaulting to: default.") });
        }
        deploySettings.nodeEnv = nodeEnv;
    }
}
exports.default = NanoDeploy;
NanoDeploy.description = 'Plateau-Symphony nano command to deploy the all functions in your app';
NanoDeploy.examples = [
    'symphony-cli sym-nano-deploy',
];
NanoDeploy.flags = {
    node_env: command_1.flags.string({
        description: 'symphony environment name',
        hidden: false,
        multiple: false,
        default: undefined,
        required: false,
        char: 'e',
    }),
    imagename: command_1.flags.string({
        description: 'image name for docker',
        hidden: false,
        multiple: false,
        default: undefined,
        required: true,
        char: 'i',
    }),
};
