// @stechquick/symphony-cli/lib/commands/sym-start.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const npmHelper_1 = require("../helpers/npmHelper");
class Run extends command_1.Command {
    async run() {
        const { flags } = this.parse(Run);
        const runSetting = {
            debug: flags.debug,
            build: flags.build,
            httpPort: flags.http_port,
            httpsPort: flags.https_port,
            hostedPath: flags.hosted_path,
            local_js_dist_path: flags.local_js_dist_path,
        };
        npmHelper_1.NpmHelper.run(flags.environment, runSetting);
    }
}
exports.default = Run;
Run.description = 'Plateau-Symphony command to start your functions with/without debug';
Run.examples = [
    `$ symphony-cli sym-start -b -d`,
    `$ symphony-cli sym-start -l ./dist/api/local.js`,
];
Run.flags = {
    debug: command_1.flags.boolean({ char: 'd', description: 'for debug' }),
    build: command_1.flags.boolean({ char: 'b', description: 'for build' }),
    http_port: command_1.flags.integer({ char: 'p', description: 'port to HTTP server' }),
    https_port: command_1.flags.integer({ char: 'h', description: 'port to HTTPS server' }),
    environment: command_1.flags.string({ char: 'e', description: 'environment variable', hidden: false, default: undefined, required: false, }),
    hosted_path: command_1.flags.string({ description: 'application hosted sub path', required: false }),
    local_js_dist_path: command_1.flags.string({ char: "l", description: "for monorepos, if local.js is not directly under dist. like: ./dist/api/local.js" }),
};
