// @stechquick/symphony-cli/lib/commands/sym-create-module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const initHelper_1 = require("../helpers/initHelper");
const chalk_1 = require("../helpers/chalk");
const log_1 = require("@stechquick/symphony-common/lib/log");
class CreateModule extends command_1.Command {
    async run() {
        const { flags } = this.parse(CreateModule);
        const { args } = this.parse(CreateModule);
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Module " + args["name"] + " creating. Please wait...") });
        await new initHelper_1.InitHelper().AddModule(args["name"], `${flags.desc}`);
        //NpmHelper.install({ folderName: args["name"] }); // Banka networkünden dolayı banka nexus olmayınca install yapılamıyor. 
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Module " + args["name"] + " created successfully.") });
    }
}
exports.default = CreateModule;
CreateModule.description = 'Plateau-Symphony command to create a Symphony Library (Module)';
CreateModule.examples = [
    `symphony-cli sym-create-module abm -d="abm module description"`
];
CreateModule.args = [
    {
        name: 'name',
        required: true,
        description: 'input module name',
    }
];
CreateModule.flags = {
    desc: command_1.flags.string({
        char: 'd',
        description: 'description to module',
        hidden: false,
        multiple: false,
        default: '',
        required: false,
    }),
};
