// @stechquick/symphony-cli/lib/helpers/ui/createuiHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateUIHelper = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const cli_ux_1 = tslib_1.__importDefault(require("cli-ux"));
const fileHelper_1 = require("@stechquick/symphony-common/lib/helpers/fileHelper");
const log_1 = require("@stechquick/symphony-common/lib/log");
const chalk_1 = require("../../helpers/chalk");
class CreateUIHelper {
    constructor() {
        this.subPath = "./";
        this.replaceMap = new Map();
        this.uiNameRegExp = /{{UIName}}/g;
        this.uiDescRegExp = /{{UIDesc}}/g;
        this.removeRegExp = /{{remove}}/g;
        this.npmrcTokenRegExp = /{{NpmrcStechQuickToken}}/g;
        this.createuiSearchPath = "../../../../templates/createui";
        this.replaceMap.set(this.removeRegExp, "");
    }
    async CreateUI(uiName, uiDesc, npmToken, registry) {
        if (npmToken == "" && registry == "") {
            log_1.Logger.log({ message: chalk_1.Chalk.create("yellow.bold", "ERR! token or registry required. Let's follow the documentation step by step: https://docs.onplateau.com/Export/Uncompiled-Export/create-ui"), timeFormat: "datetime" });
            return false;
        }
        this.replaceMap.set(this.uiNameRegExp, uiName);
        this.replaceMap.set(this.uiDescRegExp, uiDesc);
        let tokenInfo = "";
        if (npmToken) {
            tokenInfo = "@stechquick:registry=https://registry.npmjs.org/" + "\r\n" + "//registry.npmjs.org/:_authToken=" + npmToken;
        }
        else if (registry) {
            //https://scorenexus.isbank/repository/npm-stechquick-registry-proxy-repo/
            tokenInfo = "registry=" + registry + "\r\n" + "strict-ssl=false";
        }
        this.replaceMap.set(this.npmrcTokenRegExp, tokenInfo);
        const existsSubPath = fs.existsSync(this.subPath + uiName);
        if (existsSubPath) {
            const answer = await cli_ux_1.default.confirm("App: " + uiName + " already exists. If you have any, you will lose your pending changes. Would you like to overwrite it?");
            if (!answer) {
                return false;
            }
        }
        const templatePath = path.join(path.resolve(__dirname), "../.." + this.createuiSearchPath);
        return fileHelper_1.FileHelper.exportFiles({ sourcePath: templatePath, targetPath: "./", replaceMap: this.replaceMap });
    }
}
exports.CreateUIHelper = CreateUIHelper;
