// @stechquick/symphony-cli/lib/helpers/ui/quiBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FolderCopier = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const js_yaml_1 = tslib_1.__importDefault(require("js-yaml"));
const log_1 = require("@stechquick/symphony-common/lib/log");
const chalk_1 = require("../chalk");
const fileHelper_1 = require("@stechquick/algae/lib/helpers/fileHelper");
const qJsonParser_1 = require("@stechquick/algae/lib/helpers/qJsonParser");
const quickSettingsBundler_1 = require("@stechquick/algae/lib/quick/quickSettingsBundler");
const qjsonVersionHelper_1 = require("@stechquick/algae/lib/helpers/qjsonVersionHelper");
class FolderCopier {
    checkContainerServicesFile(settings) {
        try {
            if (!settings["containerServicesPath"]) {
                var containerServicesFile = fs.readFileSync("./settings/js/containerServices.ts");
            }
        }
        catch (error) {
            // let data = fs.readFileSync("../../../templates/createui/{{UIName}}/settings/js/containerServices.ts", "utf8");
            let data = fs.readFileSync("./node_modules/@stechquick/symphony-cli/templates/createui/{{UIName}}/settings/js/containerServices.ts", "utf8");
            if (!fs.existsSync("./settings/js")) {
                fs.mkdirSync("./settings/js");
            }
            fs.writeFileSync("./settings/js/containerServices.ts", data, 'utf8');
        }
    }
    async Execute() {
        fileHelper_1.DeleteFolder("./dist"); //deletes dist folder of project
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Folders copying..."), timeFormat: "datetime" });
        fileHelper_1.CopyFolder("./node_modules/@stechquick/qui/dist", "./dist", ["build"]); // copies /node_modules/@stechquick/qui/dist folder to dist
        let settingsContent = fs.readFileSync(path.resolve("./settings/settings.yaml"), { encoding: "utf-8" }); // reads settings.yaml
        let settings = js_yaml_1.default.load(settingsContent); // converts yaml file content to js
        //this.checkContainerServicesFile(settings);
        fileHelper_1.CopyFolder("./static", "./dist/client/static", undefined, {
            distinctFileExtensions: [".qjson"],
            distinctFilesCb: (srcPath) => {
                let jsonBody = fs.readFileSync(srcPath, { encoding: "utf-8" });
                jsonBody = new qJsonParser_1.QJsonParser().minifyQjson(jsonBody, settings.keepSourceMappingURL === true);
                return { success: true, value: jsonBody };
            },
            context: this
        }); // copies static folder. if encountered any file with ".qjson" extension, calls minifyQjson
        fileHelper_1.CopyFolder("./settings", "./dist/client/settings", ["server"]); // copies settings folder to dist/client/settings
        fileHelper_1.CopyFolder("./settings", "./dist/server/settings"); // copies settings folder to dist/server/settings
        qjsonVersionHelper_1.QjsonVersionHelper.mapDirectorytoSettingsYaml(settings, { safeDump: (jsObject) => js_yaml_1.default.safeDump(jsObject) });
        fileHelper_1.CopyFile("./worker/a9e657cbf4abb977edda.worker.js", "./dist/client", { copyIfExists: true }); // copies worker.js file to dist/client
        await new quickSettingsBundler_1.QuickSettingsBundler().bundleSettings("./dist", { yaml: { load: yamlContent => js_yaml_1.default.load(yamlContent) } });
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Folders copied successfully."), timeFormat: "datetime" });
    }
}
exports.FolderCopier = FolderCopier;
