// @stechquick/symphony-nanoserver-wsplugin/dist/src/nanoSocketServer.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NanoSocketServer = void 0;
var ws = __importStar(require("ws"));
var nanoSocket_1 = require("./nanoSocket");
var promiseData_1 = require("@stechquick/symphony-common/lib/helpers/promiseData");
var NanoSocketServer = /** @class */ (function () {
    function NanoSocketServer(options) {
        if (options === void 0) { options = {}; }
        this.PluginName = "wsplugin";
        this.idIndexedSockets = {};
        this.options = options;
    }
    Object.defineProperty(NanoSocketServer.prototype, "onClose", {
        set: function (cb) {
            this.options.onClose = cb;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NanoSocketServer.prototype, "onConnection", {
        set: function (cb) {
            this.options.onConnection = cb;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NanoSocketServer.prototype, "onError", {
        set: function (cb) {
            this.options.onError = cb;
        },
        enumerable: false,
        configurable: true
    });
    NanoSocketServer.prototype.Close = function () {
        if (this.wss) {
            var optionsBackup = this.options;
            this.options = {};
            this.wss.close();
            this.wss.removeAllListeners();
            this.options = optionsBackup;
        }
    };
    NanoSocketServer.prototype.Attach = function (server) {
        this.Close();
        this.attachPromise = promiseData_1.CreatePromiseData();
        this.wss = new ws.Server({ server: server, maxPayload: 500 * 1024 * 1024 });
        this.wss.on("listening", this.onSocketListening.bind(this));
        this.wss.on("connection", this.onSocketConnection.bind(this));
        this.wss.on("error", this.onSocketError.bind(this));
        this.wss.on("close", this.onSocketClose.bind(this));
    };
    NanoSocketServer.prototype.GetServerUrl = function () {
        var _a;
        var address = (_a = this.wss) === null || _a === void 0 ? void 0 : _a.address();
        if (!address) {
            return undefined;
        }
        if (typeof address == "string") {
            return address;
        }
        var protocol = false ? "wss" : "ws";
        var port = address === null || address === void 0 ? void 0 : address.port;
        var logHost = address.address == "0.0.0.0" ? "localhost" : address.address;
        return protocol + "://" + logHost + ":" + port;
    };
    NanoSocketServer.prototype.GetSocket = function (nanoSocketID) {
        return this.idIndexedSockets[nanoSocketID];
    };
    NanoSocketServer.prototype.onSocketConnection = function (ws) {
        var _this = this;
        var _a, _b;
        var cb;
        var ns = new nanoSocket_1.NanoSocket(ws, { onClose: function () { delete _this.idIndexedSockets[ns.ID]; cb === null || cb === void 0 ? void 0 : cb(); } });
        Object.defineProperty(ns, "onClose", { set: function (value) { return cb = value; }, configurable: true });
        this.idIndexedSockets[ns.ID] = ns;
        console.log("New socket conenction. Socket Id:", ns.ID);
        (_b = (_a = this.options) === null || _a === void 0 ? void 0 : _a.onConnection) === null || _b === void 0 ? void 0 : _b.call(_a, ns);
    };
    NanoSocketServer.prototype.onSocketClose = function () {
        var _a, _b;
        (_b = (_a = this.options) === null || _a === void 0 ? void 0 : _a.onClose) === null || _b === void 0 ? void 0 : _b.call(_a);
    };
    NanoSocketServer.prototype.onSocketError = function (error) {
        var _a, _b;
        if (error.stack && error.stack.indexOf("EADDRINUSE") > -1) {
            this.attachPromise.resolver({ result: false });
            return;
        }
        (_b = (_a = this.options) === null || _a === void 0 ? void 0 : _a.onError) === null || _b === void 0 ? void 0 : _b.call(_a, error);
    };
    NanoSocketServer.prototype.onSocketListening = function () {
        this.attachPromise.resolver({ result: true });
    };
    return NanoSocketServer;
}());
exports.NanoSocketServer = NanoSocketServer;
