// @stechquick/symphony-nanoserver-wsplugin/dist/src/nanoSocket.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.NanoSocket = void 0;
var cryptoHelper_1 = require("@stechquick/symphony-common/lib/helpers/cryptoHelper");
;
var NanoSocket = /** @class */ (function () {
    function NanoSocket(ws, options) {
        if (options === void 0) { options = {}; }
        this.options = options;
        this.ws = ws;
        this.ws.on("close", this.onSocketClose.bind(this));
        this.ws.on("message", this.onSocketMessage.bind(this));
        this.ws.on("error", this.onSocketError.bind(this));
        this.ws.on("open", this.onSocketOpen.bind(this));
        this.ws.on("ping", this.onSocketPing.bind(this));
        this.ws.on("pong", this.onSocketPong.bind(this));
        this.ws.on("unexpected-response", this.onSocketUnexpectedResp.bind(this));
        this.ws.on("upgrade", this.onSocketUpgrade.bind(this));
        this.id = "NS" + NanoSocket.nsCounter++;
        this.uid = cryptoHelper_1.CryptoHelper.CreateGuid();
    }
    Object.defineProperty(NanoSocket.prototype, "OnMessage", {
        get: function () {
            return this.options.onMessage;
        },
        set: function (cb) {
            this.options.onMessage = cb;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NanoSocket.prototype, "OnClose", {
        get: function () {
            return this.options.onClose;
        },
        set: function (cb) {
            this.options.onClose = cb;
        },
        enumerable: false,
        configurable: true
    });
    NanoSocket.prototype.destroy = function (code, reason) {
        this.Close(code, reason);
    };
    NanoSocket.prototype.Send = function (message) {
        var data = message instanceof Object ? JSON.stringify(message) : message;
        this.ws.send(data);
        console.log("server --> ", data.length > 50 ? data.substring(0, 50) + "... (" + data.length + ")" : data);
    };
    NanoSocket.prototype.Close = function (code, reason) {
        this.ws.close(code, reason);
    };
    Object.defineProperty(NanoSocket.prototype, "ID", {
        get: function () {
            return this.id;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(NanoSocket.prototype, "UID", {
        get: function () {
            return this.uid;
        },
        enumerable: false,
        configurable: true
    });
    NanoSocket.prototype.onSocketMessage = function (data) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_c) {
                if (typeof data == "string") {
                    console.log("server <-- ", data.length > 50 ? data.substring(0, 50) + "... (" + data.length + ")" : data);
                }
                else if (data instanceof Buffer) {
                    console.log("server <-- ... (" + data.length + ")");
                }
                else if (data instanceof Array) {
                    console.log("server <-- ... (" + data.length + ")");
                }
                else {
                    console.log("server <-- ... (" + data.byteLength + ")");
                }
                (_b = (_a = this.options).onMessage) === null || _b === void 0 ? void 0 : _b.call(_a, data);
                return [2 /*return*/];
            });
        });
    };
    NanoSocket.prototype.onSocketUnexpectedResp = function (req, resp) {
        return console.log("client sent unexpResp: ");
    };
    NanoSocket.prototype.onSocketUpgrade = function (req) {
        return console.log("client sent upgrade");
    };
    NanoSocket.prototype.onSocketPong = function (data) {
        return console.log("Client sent pong. Client Id: " + this.ID + " Data: " + data);
    };
    NanoSocket.prototype.onSocketPing = function (data) {
        return console.log("client sent ping: " + data);
    };
    NanoSocket.prototype.onSocketOpen = function () {
        console.log("client connected: " + this.ID);
    };
    NanoSocket.prototype.onSocketError = function (err) {
        console.log("client err " + this.ID + ": ", err);
    };
    NanoSocket.prototype.onSocketClose = function (code, reason) {
        var _a, _b;
        this.ws.removeAllListeners();
        (_b = (_a = this.options) === null || _a === void 0 ? void 0 : _a.onClose) === null || _b === void 0 ? void 0 : _b.call(_a);
        console.log("client connection close: " + this.ID + " code: ", code, "reason: ", reason);
    };
    NanoSocket.nsCounter = 0;
    return NanoSocket;
}());
exports.NanoSocket = NanoSocket;
