// @stechquick/symphony-common/lib/helpers/arrayHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArrayHelper = void 0;
class ArrayHelper {
    static findAndRet(array, predicate) {
        const len = array.length;
        for (var i = 0; i < len; i++) {
            const retVal = predicate(array[i], i, array);
            if (retVal !== undefined) {
                return retVal;
            }
        }
        return undefined;
    }
    static InitArray(size) {
        const arr = [];
        for (let i = 0; i < size; i++) {
            arr.push(i);
        }
        return arr;
    }
    static InitEmptyArray(size) {
        return Array.from(new Array(size));
    }
    /**
     * Fast way to find last object in a given array
     * @param array to be search in
     * @param key to be last searched  object key
     */
    static lastIndexOf(array, key, value) {
        for (let i = array.length - 1; i >= 0; i--) {
            if (array[i][key] === (value ? value : key))
                return i;
        }
        return -1;
    }
    ;
    /**
     * Change object of index
     * First index can move to the end of the array
     * Last index can move to first index
     * @param array Array to be moved
     * @param oldIndex Old index of object
     * @param newIndex New index of object
     */
    static arrayMove(array, oldIndex, newIndex) {
        if (newIndex < 0) {
            array.splice(array.length - 1, 0, array.splice(oldIndex, 1)[0]);
        }
        else if (newIndex == array.length) {
            array.splice(0, 0, array.splice(oldIndex, 1)[0]);
        }
        else {
            array.splice(newIndex, 0, array.splice(oldIndex, 1)[0]);
        }
    }
    ;
}
exports.ArrayHelper = ArrayHelper;
