// @stechquick/symphony-common/lib/helpers/fileHelper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileHelper = exports.ItemType = void 0;
const fs = __importStar(require("fs"));
const path = __importStar(require("path"));
const logger_1 = require("./logger");
const ILogItem_1 = require("../interfaces/ILogItem");
const colors_1 = require("./colors");
var ItemType;
(function (ItemType) {
    ItemType[ItemType["Directory"] = 0] = "Directory";
    ItemType[ItemType["File"] = 1] = "File";
    ItemType[ItemType["FIFO"] = 2] = "FIFO";
    ItemType[ItemType["Socket"] = 3] = "Socket";
    ItemType[ItemType["SymbolicLink"] = 4] = "SymbolicLink";
})(ItemType = exports.ItemType || (exports.ItemType = {}));
class FileHelper {
    static get Slash() {
        return FileHelper.slashCache || (FileHelper.slashCache = path.resolve("/") == path.posix.resolve("/") ? "/" : "\\");
    }
    static removeDist() {
        this.deleteFolderRecursive({ path: "./dist" });
    }
    static removeFolder(path) {
        this.deleteFolderRecursive({ path });
    }
    static deleteFolderRecursive({ path, deleteRootPath }) {
        logger_1.Logger.log({ message: "Dist file cleaning is started...", level: ILogItem_1.LogLevel.log });
        const rootPath = path;
        const deleteRecursiveFunc = (path) => {
            if (!fs.existsSync(path)) {
                //Logger.log({ message: "The path not exists: " + path, level: LogLevel.log });
                return;
            }
            fs.readdirSync(path).forEach(function (file, index) {
                var curPath = path + "/" + file;
                if (fs.lstatSync(curPath).isDirectory()) {
                    deleteRecursiveFunc(curPath);
                }
                else {
                    //Logger.log({ message: "Deleting item: " + curPath, level: LogLevel.log });
                    fs.unlinkSync(curPath);
                }
            });
            if (!deleteRootPath && rootPath == path) {
                //Logger.log({ message: "Passing the root path...", level: LogLevel.log });
                return;
            }
            //Logger.log({ message: "Deleting item: " + path, level: LogLevel.log });
            fs.rmdirSync(path);
        };
        deleteRecursiveFunc(path);
        logger_1.Logger.log({ message: "Dist file cleaning is finished...", level: ILogItem_1.LogLevel.log });
    }
    static exportFilesInner(settings) {
        fs.readdirSync(settings.sourcePath).forEach(item => {
            let itemReplaced = item;
            settings.replaceMap.forEach(function (newstring, old) {
                itemReplaced = itemReplaced.replace(old, newstring);
            });
            const sourceItem = path.join(settings.sourcePath, item);
            const targetItem = path.join(settings.targetPath, itemReplaced);
            const data = fs.lstatSync(sourceItem);
            if (data.isDirectory()) {
                fs.mkdirSync(targetItem, { recursive: true });
                //Logger.log({ message: `${targetItem} is saved.`});
                this.exportFilesInner({ sourcePath: sourceItem, targetPath: targetItem, replaceMap: settings.replaceMap });
            }
            else {
                const ext = path.extname(sourceItem);
                if (this.discardReplacementExt.indexOf(ext) > -1) {
                    logger_1.Logger.log({ message: `${colors_1.colors.green}${colors_1.colors.bold}Replacement skipping for ${targetItem}${colors_1.colors.reset}` });
                    const data = fs.readFileSync(sourceItem);
                    fs.writeFileSync(targetItem, data);
                    return;
                }
                let data = fs.readFileSync(sourceItem, "utf8");
                settings.replaceMap.forEach(function (newstring, old) {
                    data = data.replace(old, newstring);
                });
                fs.writeFileSync(targetItem, data, 'utf8');
                //Logger.log({ message: `${targetItem} is saved.` });
            }
        });
    }
    static exportFiles(settings) {
        logger_1.Logger.log({ message: `${colors_1.colors.green}${colors_1.colors.bold}Files copying...${colors_1.colors.reset}` });
        this.exportFilesInner(settings);
        logger_1.Logger.log({ message: `${colors_1.colors.green}${colors_1.colors.bold}Files copied successfully.${colors_1.colors.reset}` });
    }
    static getFileNamesFormPath({ selectedPath, excludeFolders, selectedExtensions }) {
        let retVal = [];
        const isSelectedExtension = (extension, selectedExtensions) => {
            if (!selectedExtensions) {
                return true; //Özellikle seçilecek extension belirtmediyse hepsini listeye ekleyebiliriz.
            }
            return selectedExtensions.indexOf(extension) > -1
                || selectedExtensions.indexOf(extension.substring(1, extension.length)) > -1;
        };
        const files = fs.readdirSync(path.resolve(selectedPath));
        files.forEach(name => {
            const ext = path.extname(name);
            const isDirectory = this.isDirectory(path.join(selectedPath, name));
            //Folder ignore check
            if (isDirectory && !excludeFolders) {
                retVal.push({ name, type: "folder" });
            }
            //Selected extensions check
            else if (isSelectedExtension(ext, selectedExtensions)) {
                retVal.push({ name: path.basename(name, ext), ext, type: "file" });
            }
        });
        return retVal;
    }
    static isDirectory(itemPath) {
        // Why try-catch?: https://stackoverflow.com/a/33401005
        try {
            const stat = fs.statSync(itemPath);
            return stat.isDirectory();
        }
        catch {
            return false;
        }
    }
}
exports.FileHelper = FileHelper;
FileHelper.discardReplacementExt = [".jar"];
