// @stechquick/symphony-common/lib/helpers/typeHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TypeHelper = void 0;
class TypeHelper {
    isJsonString(param) {
        try {
            JSON.parse(param);
        }
        catch (e) {
            return false;
        }
        return true;
    }
    GetObjectType(param) {
        return Object.prototype.toString.call(param);
    }
    ;
    /** @deprecated use instanceof for typescript type distinction intellisense. ex: if (variable instanceof Array)*/
    isArray(param) {
        return this.GetObjectType(param) == "[object Array]";
    }
    ;
    isString(param) {
        return this.GetObjectType(param) === "[object String]";
    }
    ;
    isNumber(param) {
        return this.GetObjectType(param) === "[object Number]";
    }
    ;
    isNull(param) {
        return param === null || param === undefined;
    }
    ;
    isFunction(param) {
        return this.GetObjectType(param) === "[object Function]";
    }
    ;
    isObject(param) {
        return this.GetObjectType(param) === "[object Object]";
    }
    ;
    isBool(param) {
        return this.GetObjectType(param) === "[object Boolean]";
    }
    ;
    isBooleanConstructor(param) {
        return (param === null || param === void 0 ? void 0 : param.toString()) == "function Boolean() { [native code] }";
    }
    isStringConstructor(param) {
        return (param === null || param === void 0 ? void 0 : param.toString()) == "function String() { [native code] }";
    }
    isArrayConstructor(param) {
        return (param === null || param === void 0 ? void 0 : param.toString()) == "function Array() { [native code] }";
    }
    isObjectConstructor(param) {
        return (param === null || param === void 0 ? void 0 : param.toString()) == "function Object() { [native code] }";
    }
    isNumberConstructor(param) {
        return (param === null || param === void 0 ? void 0 : param.toString()) == "function Number() { [native code] }";
    }
    isDateString(param) {
        if (!this.isString(param)) {
            return false;
        }
        return !isNaN(Date.parse(param));
    }
    ;
    isDate(param) {
        return this.GetObjectType(param) === "[object Date]";
    }
    isFile(param) {
        return this.GetObjectType(param) === "[object File]";
    }
    cloneObject(obj) {
        var copy;
        // Handle the 3 simple types, and null or undefined
        if (null == obj || "object" != typeof obj) {
            return obj;
        }
        // Handle Date
        if (obj instanceof Date) {
            copy = new Date();
            copy.setTime(obj.getTime());
            return copy;
        }
        // Handle Array
        if (Array.isArray(obj)) {
            copy = [];
            for (var i = 0, len = obj.length; i < len; i++) {
                copy[i] = this.cloneObject(obj[i]);
            }
            return copy;
        }
        // Handle Object
        if (typeof obj === "object") {
            copy = obj.constructor && obj.constructor.length < 1 ? new obj.constructor() : {};
            for (var attr in obj) {
                if (obj.hasOwnProperty(attr)) {
                    copy[attr] = this.cloneObject(obj[attr]);
                }
            }
            return copy;
        }
        throw new Error("Unable to copy obj! Its type isn't supported.");
    }
    ;
}
exports.TypeHelper = TypeHelper;
