// @stechquick/symphony-cli/lib/predeploy/src/predeployCommon.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PredeployCommon = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const yaml = tslib_1.__importStar(require("js-yaml"));
const fileHelper_1 = require("@stechquick/symphony-common/lib/helpers/fileHelper");
const paramReplacer_1 = require("./paramReplacer");
const fileHelper_2 = require("@stechquick/algae/lib/helpers/fileHelper");
class PredeployCommon {
    constructor() {
        this.config = {};
    }
    forEachFolderWithItem(options) {
        const rootFolder = options.folder;
        ////logger.log({ message: "Root folder: " + rootFolder });
        let funcName;
        const forEachFolder = ({ options }) => {
            fs.readdirSync(options.folder).forEach(item => {
                ////logger.log({ message: "Item: " + item });
                const curItemPath = path.join(options.folder, item);
                ////logger.log({ message: "Current item path: " + curItemPath });
                const isDirectory = fileHelper_1.FileHelper.isDirectory(curItemPath);
                ////logger.log({ message: "Now, we are in a directory?: " + isDirectory });
                if (isDirectory) {
                    options.folder = curItemPath;
                    funcName = item;
                    forEachFolder({ options });
                }
                else if (item.endsWith(options.itemEndsWith)) {
                    const writeFile = (config) => {
                        config["name"] = this.config["name"].toLowerCase();
                        fs.writeFileSync(curItemPath, yaml.safeDump(config), 'utf8');
                    };
                    let funcYaml = yaml.safeLoad(fs.readFileSync(curItemPath, "utf8"));
                    switch (typeof (funcYaml)) {
                        case "string":
                            this.config = JSON.parse(funcYaml);
                            writeFile(this.config);
                            break;
                        case "object":
                            this.config = funcYaml;
                            writeFile(this.config);
                        case "undefined":
                            break;
                    }
                    const iteratingItem = curItemPath.substring(0, curItemPath.length - options.itemEndsWith.length);
                    ////logger.log({ message: "Iterating item: " + iteratingItem });
                    options.iterator(iteratingItem, funcName);
                }
            });
            //Set for next function folder
            options.folder = rootFolder;
        };
        forEachFolder({ options });
    }
    preDeployReplaceCopier(targetPath, funcName, replaceWords, copyList) {
        // //logger.log({ message: "I am in iterator... targetPath: " + targetPath });
        copyList.forEach(copyItem => {
            const parsedCopyItem = path.parse(copyItem.path);
            copyItem.path = copyItem.sourceFullPath ? copyItem.path : `./${copyItem.path}`;
            const targetItem = targetPath + parsedCopyItem.name + parsedCopyItem.ext;
            ////logger.log({ message: "From: " + copyItem.path + " To: " + targetItem });
            if (copyItem.needParamReplace) {
                const content = fs.readFileSync(copyItem.path, "utf8");
                replaceWords.funcName = funcName;
                const newContent = paramReplacer_1.ParamReplacer.replace({ content, replaceWords });
                fs.writeFileSync(targetItem, newContent);
            }
            else {
                if (fileHelper_1.FileHelper.isDirectory(copyItem.path)) {
                    fileHelper_2.CopyFolder(copyItem.path, targetItem);
                }
                else {
                    fs.copyFileSync(copyItem.path, targetItem);
                }
            }
            ////logger.log({ message: "item was coppied... copyItem: " + copyItem.path + " targetItem: " + targetItem });
        });
    }
}
exports.PredeployCommon = PredeployCommon;
