// @stechquick/algae/lib/helpers/eventBus.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventBus = void 0;
var hook_1 = require("./hook");
var EventBus = /** @class */ (function () {
    function EventBus() {
        this.subsDict = {};
    }
    EventBus.prototype.obtainSubsList = function (eventName) {
        return this.subsDict[eventName] || (this.subsDict[eventName] = new hook_1.Hook("EventBus_" + eventName));
    };
    EventBus.prototype.Subscribe = function (eventName, cb) {
        if (eventName == EventBus.AllEvents) {
            throw new Error(eventName + " is a registered key. If you want to subscribe all events, use SubsribeAll().");
        }
        this.obtainSubsList(eventName).subscribe(cb);
    };
    EventBus.prototype.Unsubscribe = function (eventName, cb) {
        var _a;
        if (eventName == EventBus.AllEvents) {
            throw new Error(eventName + " is a registered key. If you want to subscribe all events, use SubsribeAll().");
        }
        (_a = this.subsDict[eventName]) === null || _a === void 0 ? void 0 : _a.unsubscribe(cb);
    };
    EventBus.prototype.SubscribeAll = function (cb) {
        this.obtainSubsList(EventBus.AllEvents).subscribe(cb);
    };
    EventBus.prototype.UnsubscribeAll = function (cb) {
        var _a;
        (_a = this.subsDict[EventBus.AllEvents]) === null || _a === void 0 ? void 0 : _a.unsubscribe(cb);
    };
    EventBus.prototype.Trigger = function (eventName) {
        var params = [];
        for (var _i = 1; _i < arguments.length; _i++) {
            params[_i - 1] = arguments[_i];
        }
        var allEventsHook = this.subsDict[EventBus.AllEvents];
        allEventsHook === null || allEventsHook === void 0 ? void 0 : allEventsHook.triggerer.apply(allEventsHook, params);
        if (eventName == EventBus.AllEvents) {
            return;
        }
        var hook = this.subsDict[eventName];
        hook === null || hook === void 0 ? void 0 : hook.triggerer.apply(hook, params);
    };
    EventBus.prototype.TriggerJustAllSubs = function () {
        var params = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            params[_i] = arguments[_i];
        }
        var allEventsHook = this.subsDict[EventBus.AllEvents];
        allEventsHook === null || allEventsHook === void 0 ? void 0 : allEventsHook.triggerer.apply(allEventsHook, params);
    };
    EventBus.prototype.destroy = function () {
        this.subsDict = {};
    };
    EventBus.AllEvents = "AllEvents";
    return EventBus;
}());
exports.EventBus = EventBus;
