// @stechquick/algae/lib/helpers/executionHelper.js
"use strict";
/**
 * Action'ların key üzerinden ayrılması.
 *  Key <TKey> olabilir, Record<string, string> -> obje alanları concat ederek string key elde edilir, adama verirken TKey'den verilir
 * Backend timing'in client performans alanına yazdırılması
 * capture | mock | off from terminal
 * Action'lara cache setting
 * Action'ların pre ve post'u
 * Sym-Lib üzerinden Pipeline'a tüm actionlar için genel pre ve post. Actionlar TKey ile merkezi ayrıştırılabilir.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Execution = void 0;
var Action = /** @class */ (function () {
    function Action(key, options) {
        this.key = key;
        this.options = options;
        this.parts = [];
    }
    Object.defineProperty(Action.prototype, "isStarted", {
        get: function () {
            var last = this.getLast();
            if (!last || !last.endTime) {
                return false;
            }
            return true;
        },
        enumerable: false,
        configurable: true
    });
    Action.prototype.getFirst = function () { return this.parts[0]; };
    Action.prototype.getLast = function () { return this.parts[this.parts.length - 1]; };
    Action.prototype.scope = function (cb) {
        return __awaiter(this, void 0, void 0, function () {
            var retVal, error_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.start();
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, cb()];
                    case 2:
                        retVal = _a.sent();
                        this.end();
                        return [2 /*return*/, retVal];
                    case 3:
                        error_1 = _a.sent();
                        this.end({ error: error_1 });
                        throw error_1;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    Action.prototype.start = function () {
        var last = this.getLast();
        if (!last || last.endTime) {
            this.parts.push({ startTime: Date.now(), startCounter: 1 });
            return;
        }
        if (!this.options.enableParalelStart) {
            throw new Error("already started: " + this.key);
        }
        last.startCounter++;
    };
    Action.prototype.end = function (options) {
        var last = this.getLast();
        if (!last) {
            throw new Error("not yet started: " + this.key);
        }
        --last.startCounter == 0 && (last.endTime = Date.now());
        if (last.startCounter < 0) {
            throw new Error("already ended: " + this.key + " counter: " + last.startCounter);
        }
        (options === null || options === void 0 ? void 0 : options.error) && (last.error = options === null || options === void 0 ? void 0 : options.error);
    };
    Action.prototype.report = function () {
        var last = this.getLast();
        var first = this.getFirst();
        return {
            key: this.key,
            startTime: (first === null || first === void 0 ? void 0 : first.startTime) || -1,
            total: this.parts.reduce(function (prevSum, cur) { return prevSum + (cur.endTime || Date.now()) - cur.startTime; }, 0),
            success: this.parts.find(function (park) { return park.error; }) == undefined,
            finished: (last === null || last === void 0 ? void 0 : last.endTime) != undefined,
        };
    };
    return Action;
}());
var Execution = /** @class */ (function () {
    function Execution(key) {
        this.key = key;
        this.startTime = Date.now();
        this.actions = [];
        this.counterDict = {};
    }
    Execution.startExecutionScope = function (key, cb) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, new Execution(key).scope(cb)];
                    case 1: return [2 /*return*/, _a.sent()];
                }
            });
        });
    };
    Execution.prototype.assertNotEnded = function () { if (this.endTime) {
        throw new Error("execution already ended");
    } };
    Execution.prototype.scope = function (cb) {
        return __awaiter(this, void 0, void 0, function () {
            var retVal, error_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.assertNotEnded();
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, cb(this)];
                    case 2:
                        retVal = _a.sent();
                        this.end();
                        return [2 /*return*/, retVal];
                    case 3:
                        error_2 = _a.sent();
                        this.end({ error: error_2 });
                        throw error_2;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    Execution.prototype.end = function (options) {
        this.assertNotEnded();
        this.error = options === null || options === void 0 ? void 0 : options.error;
        this.endTime = Date.now();
    };
    Execution.prototype.startActionScope = function (options, cb) {
        return __awaiter(this, void 0, void 0, function () {
            var action, retVal, error_3;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        action = this.startAction(options.key, { transient: options.transient, appendCounterToKey: options.appendCounterToKey });
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, cb()];
                    case 2:
                        retVal = _a.sent();
                        action.end();
                        return [2 /*return*/, retVal];
                    case 3:
                        error_3 = _a.sent();
                        action.end({ error: error_3 });
                        throw error_3;
                    case 4: return [2 /*return*/];
                }
            });
        });
    };
    /**
     *
     * @param options {transient: action will not be reported, appendCounterToKey: manage an auto incremented counter for key and append it to key, enableParalelStart: enable multiple start calls}
     */
    Execution.prototype.startAction = function (key, options) {
        var _a = options || {}, enableParalelStart = _a.enableParalelStart, appendCounterToKey = _a.appendCounterToKey;
        var action = this.createAction(key, { appendCounterToKey: appendCounterToKey, enableParalelStart: enableParalelStart });
        !(options === null || options === void 0 ? void 0 : options.transient) && this.actions.push(action);
        if (!(options === null || options === void 0 ? void 0 : options.doNotStart)) {
            action.start();
        }
        return action;
    };
    Execution.prototype.createAction = function (actionKey, _a) {
        var appendCounterToKey = _a.appendCounterToKey, enableParalelStart = _a.enableParalelStart;
        if (!appendCounterToKey) {
            return new Action(actionKey, { enableParalelStart: enableParalelStart });
        }
        var sameKeyActions = this.counterDict[actionKey] = this.counterDict[actionKey] || [];
        var newKey = actionKey + "_" + sameKeyActions.length;
        var action = new Action(newKey, { enableParalelStart: enableParalelStart });
        sameKeyActions.push(action);
        return action;
    };
    Execution.prototype.report = function () {
        if (this.endReport) {
            return this.endReport;
        }
        var actionsReport = this.actions.map(function (m) { return m.report(); });
        var report = {
            key: this.key,
            success: this.error == undefined,
            total: (this.endTime || Date.now()) - this.startTime,
            startTime: this.startTime,
            actions: actionsReport,
            finished: this.endTime != undefined,
        };
        return this.endTime ? this.endReport = report : report;
    };
    return Execution;
}());
exports.Execution = Execution;
var sleep = function (time) { return new Promise(function (res) { return setTimeout(res, time); }); };
var stuff = function () { return __awaiter(void 0, void 0, void 0, function () { return __generator(this, function (_a) {
    switch (_a.label) {
        case 0:
            console.log("doing real stuff");
            return [4 /*yield*/, sleep(100)];
        case 1:
            _a.sent();
            return [2 /*return*/, "real stuff"];
    }
}); }); };
function test1() {
    return __awaiter(this, void 0, void 0, function () {
        var execution, ex_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    execution = new Execution("test1");
                    _a.label = 1;
                case 1:
                    _a.trys.push([1, 4, , 5]);
                    return [4 /*yield*/, actionTest(execution)];
                case 2:
                    _a.sent();
                    return [4 /*yield*/, doAsyncStuff(execution)];
                case 3:
                    _a.sent();
                    execution.end();
                    setTimeout(function () {
                        var act = execution.startAction("x1", { enableParalelStart: true, appendCounterToKey: true });
                        act.start();
                        act.end();
                        act.end();
                        console.log(execution.report());
                    });
                    return [3 /*break*/, 5];
                case 4:
                    ex_1 = _a.sent();
                    execution.end({ error: ex_1 });
                    return [3 /*break*/, 5];
                case 5: return [2 /*return*/];
            }
        });
    });
}
function test2() {
    return __awaiter(this, void 0, void 0, function () {
        var _this = this;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0: return [4 /*yield*/, Execution.startExecutionScope("test2", function (execution) { return __awaiter(_this, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, doAsyncStuff(execution)];
                                case 1:
                                    _a.sent();
                                    return [2 /*return*/];
                            }
                        });
                    }); })];
                case 1:
                    _a.sent();
                    return [2 /*return*/];
            }
        });
    });
}
function actionTest(execution) {
    return __awaiter(this, void 0, void 0, function () {
        var act1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    act1 = execution.startAction("x1", { appendCounterToKey: true, enableParalelStart: true, doNotStart: true });
                    act1.start();
                    return [4 /*yield*/, sleep(10)];
                case 1:
                    _a.sent();
                    act1.end();
                    act1.start();
                    return [4 /*yield*/, sleep(10)];
                case 2:
                    _a.sent();
                    act1.start();
                    return [4 /*yield*/, sleep(10)];
                case 3:
                    _a.sent();
                    act1.end();
                    return [4 /*yield*/, sleep(10)];
                case 4:
                    _a.sent();
                    act1.end();
                    console.log(act1.report());
                    return [2 /*return*/];
            }
        });
    });
}
var doAsyncStuff = function (execution) { return __awaiter(void 0, void 0, void 0, function () {
    var action, result_1, result0, result1, result2;
    return __generator(this, function (_a) {
        switch (_a.label) {
            case 0:
                action = execution.startAction("pre");
                action.end();
                return [4 /*yield*/, stuff()];
            case 1:
                result_1 = _a.sent();
                action.start();
                action.end();
                return [4 /*yield*/, stuff()];
            case 2:
                result0 = _a.sent();
                return [4 /*yield*/, execution.startActionScope({ key: "doAsyncStuff1" }, stuff)];
            case 3:
                result1 = _a.sent();
                return [4 /*yield*/, execution.startActionScope({ key: "doAsyncStuff2" }, function () { return __awaiter(void 0, void 0, void 0, function () {
                        return __generator(this, function (_a) {
                            switch (_a.label) {
                                case 0: return [4 /*yield*/, stuff()];
                                case 1: return [2 /*return*/, _a.sent()];
                            }
                        });
                    }); })];
            case 4:
                result2 = _a.sent();
                console.log(result1, result2);
                return [2 /*return*/];
        }
    });
}); };
