// @stechquick/algae/lib/helpers/fileHelper.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.exists = exports.stats = exports.readDir = exports.writeFile = exports.createFolders = exports.readFile = exports.tryGetContents = exports.zipFolder = exports.DeleteFolder = exports.CopyFile = exports.CopyFolder = void 0;
var fs = __importStar(require("fs"));
var path = __importStar(require("path"));
var fs_1 = require("fs");
var fsPromises = fs_1.promises;
function ForEachItemInFolder(folder, iterator, willIterateFolder) {
    function forEachItemInFoderInternal(folder, curRelative) {
        fs.readdirSync(folder).forEach(function (item) {
            var curItemPath = path.join(folder, item);
            var isDirectory = fs.statSync(curItemPath).isDirectory();
            if (isDirectory) {
                if (!willIterateFolder || (willIterateFolder === null || willIterateFolder === void 0 ? void 0 : willIterateFolder(folder, item, curRelative))) {
                    forEachItemInFoderInternal(curItemPath, curRelative + "/" + item);
                    iterator(folder, item, curRelative, true);
                }
            }
            else {
                iterator(folder, item, curRelative, false);
            }
        });
    }
    forEachItemInFoderInternal(folder, "");
}
;
/**
 * copies folder from source to destination. if except param is given, excepts folders or files.
 * @param source
 * @param destination
 * @param except
 */
function CopyFolder(source, destination, except, distinction) {
    var exceptDict = (except === null || except === void 0 ? void 0 : except.length) ? except.reduce(function (prev, current) { prev[current] = true; return prev; }, {}) : undefined;
    ForEachItemInFolder(source, function (folder, item, relativeFolder, isDirectory) {
        if (isDirectory) {
            return;
        }
        if (exceptDict) {
            var relativeItemPath = path.join(relativeFolder, item);
            if (exceptDict[relativeItemPath]) {
                return;
            }
        }
        var srcPath = path.resolve(path.join(folder, item));
        var destPath = path.resolve(path.join(destination, relativeFolder));
        if (!fs.existsSync(destPath)) {
            fs.mkdirSync(destPath, { recursive: true });
        }
        if (distinction === null || distinction === void 0 ? void 0 : distinction.distinctFileExtensions.includes(path.extname(item))) {
            var modifiedFileContent = distinction.distinctFilesCb.apply(distinction.context, [srcPath]);
            if (modifiedFileContent.success) {
                fs.writeFileSync(path.join(destPath, item), modifiedFileContent.value);
            }
        }
        else {
            fs.copyFileSync(srcPath, path.join(destPath, item));
        }
    }, function (folder, item, relativeFolder) {
        var relativeItemPath = path.join(relativeFolder, item);
        if (exceptDict === null || exceptDict === void 0 ? void 0 : exceptDict[relativeItemPath]) {
            return false;
        }
        return true;
    });
}
exports.CopyFolder = CopyFolder;
/**
 * copies file source to destination
 * @param source
 * @param destination
 * @param options
 * @returns
 */
function CopyFile(source, destination, options) {
    if (options === void 0) { options = { recursive: true }; }
    var pathParsed = path.parse(source);
    var srcFile = path.resolve(source);
    if (options.copyIfExists && !fs.existsSync(srcFile)) {
        return;
    }
    var destPath = path.resolve(destination);
    if (!fs.existsSync(destPath)) {
        fs.mkdirSync(destPath, { recursive: options.recursive });
    }
    var destFile = path.join(destPath, pathParsed.base);
    fs.copyFileSync(srcFile, destFile);
}
exports.CopyFile = CopyFile;
/**
 * deletes folder recursively
 * @param target
 * @returns
 */
function DeleteFolder(target) {
    if (!fs.existsSync(target)) {
        return;
    }
    ForEachItemInFolder(target, function (folder, item, relativeFolder, isDirectory) {
        var curPath = path.join(folder, item);
        if (isDirectory) {
            fs.rmdirSync(curPath);
        }
        else {
            fs.unlinkSync(path.join(folder, item));
        }
    });
}
exports.DeleteFolder = DeleteFolder;
function zipFolder(directory, outputFile, errCallback) {
    return __awaiter(this, void 0, void 0, function () {
        var _this = this;
        return __generator(this, function (_a) {
            return [2 /*return*/, new Promise(function (resolve, reject) { return __awaiter(_this, void 0, void 0, function () {
                    var output, archiver, zip, archive, error_1;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                _a.trys.push([0, 2, , 3]);
                                output = fs.createWriteStream(outputFile);
                                return [4 /*yield*/, Promise.resolve().then(function () { return __importStar(require("archiver")); })];
                            case 1:
                                archiver = _a.sent();
                                zip = archiver.default;
                                archive = zip('zip', { zlib: { level: 9 } });
                                // listen for all archive data to be written
                                // 'close' event is fired only when a file descriptor is involved
                                output.on('close', function () {
                                    resolve(outputFile);
                                });
                                output.on('error', function (err) {
                                    throw err;
                                });
                                // good practice to catch this error explicitly
                                archive.on('error', function (err) {
                                    throw err;
                                });
                                // pipe archive data to the file
                                archive.pipe(output);
                                archive.directory(directory, false);
                                archive.finalize();
                                return [3 /*break*/, 3];
                            case 2:
                                error_1 = _a.sent();
                                errCallback === null || errCallback === void 0 ? void 0 : errCallback(error_1);
                                reject(error_1);
                                return [3 /*break*/, 3];
                            case 3: return [2 /*return*/];
                        }
                    });
                }); })];
        });
    });
}
exports.zipFolder = zipFolder;
function tryGetContents(filePath, errCallback) {
    return __awaiter(this, void 0, void 0, function () {
        var retVal, err_1;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, fsPromises.readFile(filePath, { encoding: "utf-8" })];
                case 1:
                    retVal = _a.sent();
                    return [3 /*break*/, 3];
                case 2:
                    err_1 = _a.sent();
                    errCallback === null || errCallback === void 0 ? void 0 : errCallback(err_1);
                    return [3 /*break*/, 3];
                case 3: return [2 /*return*/, retVal];
            }
        });
    });
}
exports.tryGetContents = tryGetContents;
function readFile(fileName) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, new Promise(function (res, rej) {
                    fs.readFile(fileName, { encoding: "utf-8" }, function (err, data) {
                        if (err) {
                            rej(err);
                            return;
                        }
                        res(data);
                    });
                })];
        });
    });
}
exports.readFile = readFile;
function createFolders(folderPath) {
    var _this = this;
    return new Promise(function (res, rej) { return __awaiter(_this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            fs.mkdir(folderPath, { recursive: true }, function (err) {
                if (err) {
                    rej(err);
                    return;
                }
                res();
            });
            return [2 /*return*/];
        });
    }); });
}
exports.createFolders = createFolders;
function writeFile(fileName, value, _a) {
    var _b = _a === void 0 ? {} : _a, _c = _b.encoding, encoding = _c === void 0 ? "utf-8" : _c, mode = _b.mode, createFolders_param = _b.createFolders;
    return __awaiter(this, void 0, void 0, function () {
        var _this = this;
        return __generator(this, function (_d) {
            return [2 /*return*/, new Promise(function (res, rej) { return __awaiter(_this, void 0, void 0, function () {
                    var folder;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                if (!createFolders_param) return [3 /*break*/, 2];
                                folder = path.parse(fileName).dir;
                                return [4 /*yield*/, createFolders(folder)];
                            case 1:
                                _a.sent();
                                _a.label = 2;
                            case 2:
                                fs.writeFile(fileName, value, { encoding: encoding, mode: mode }, function (err) {
                                    if (err) {
                                        rej(err);
                                        return;
                                    }
                                    res();
                                });
                                return [2 /*return*/];
                        }
                    });
                }); })];
        });
    });
}
exports.writeFile = writeFile;
function readDir(folderPath) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, new Promise(function (res, rej) {
                    fs.readdir(folderPath, function (err, files) {
                        if (err) {
                            rej(err);
                            return;
                        }
                        res(files);
                    });
                })];
        });
    });
}
exports.readDir = readDir;
;
function stats(itemPath) {
    return __awaiter(this, void 0, void 0, function () {
        return __generator(this, function (_a) {
            return [2 /*return*/, new Promise(function (res, rej) {
                    fs.stat(itemPath, function (err, stats) {
                        if (err) {
                            rej(err);
                            return;
                        }
                        res(stats);
                    });
                })];
        });
    });
}
exports.stats = stats;
;
function exists(itemPath) {
    return __awaiter(this, void 0, void 0, function () {
        var err_2;
        return __generator(this, function (_a) {
            switch (_a.label) {
                case 0:
                    _a.trys.push([0, 2, , 3]);
                    return [4 /*yield*/, stats(itemPath)];
                case 1:
                    _a.sent();
                    return [2 /*return*/, true];
                case 2:
                    err_2 = _a.sent();
                    if (err_2.code === 'ENOENT') {
                        return [2 /*return*/, false];
                    }
                    ;
                    throw err_2;
                case 3: return [2 /*return*/];
            }
        });
    });
}
exports.exists = exists;
