// @stechquick/algae/lib/helpers/frameCommunicator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FrameCommunicator = void 0;
var structuredSocket_1 = require("./socket/structuredSocket");
var WindowSocket = /** @class */ (function () {
    function WindowSocket(targetWindow) {
        var _this = this;
        this.OnMessageWithDirectObject = true;
        this.targetWindow = targetWindow;
        window.addEventListener("message", this.messageHandler = function (ev) {
            var _a;
            if (ev.source != targetWindow) {
                return;
            }
            (_a = _this.OnMessage) === null || _a === void 0 ? void 0 : _a.call(_this, ev.data);
        });
    }
    WindowSocket.prototype.Send = function (packet) {
        this.targetWindow.postMessage(packet, "*");
    };
    WindowSocket.prototype.destroy = function () {
        window.removeEventListener("message", this.messageHandler);
        this.OnMessage = undefined;
    };
    return WindowSocket;
}());
var FrameCommunicator = /** @class */ (function () {
    function FrameCommunicator(targetWindow) {
        this.windowSocket = new WindowSocket(targetWindow);
        this.structuredSocket = new structuredSocket_1.StructuredSocket(this.windowSocket);
    }
    FrameCommunicator.prototype.Send = function (target, messageType, message, options) {
        var structuredSocket = this.structuredSocket;
        return structuredSocket.Send(message, options);
    };
    FrameCommunicator.prototype.Subscribe = function (target, messageType, cb) {
        var structuredSocket = this.structuredSocket;
        structuredSocket.Subscribe(messageType, cb);
    };
    FrameCommunicator.prototype.SubscribeAll = function (cb) {
        this.structuredSocket.SubscribeAll(function (data) { return cb(data); });
    };
    FrameCommunicator.prototype.UnsubscribeAll = function (cb) {
        this.structuredSocket.UnsubscribeAll(function (data) { return cb(data); });
    };
    FrameCommunicator.prototype.destroy = function () {
        this.structuredSocket.destroy();
    };
    return FrameCommunicator;
}());
exports.FrameCommunicator = FrameCommunicator;
