// @stechquick/algae/lib/helpers/qVersionManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QVersionManager = void 0;
var arrayHelper_1 = require("./arrayHelper");
var QVersionManager = /** @class */ (function () {
    function QVersionManager() {
        this.numChars = { "1": true, "2": true, "3": true, "4": true, "5": true, "6": true, "7": true, "8": true, "9": true, "0": true };
    }
    QVersionManager.prototype.extractPrependNumber = function (data) {
        var bucket = "";
        var curData;
        var i = 0;
        for (i = 0; i < data.length; i++) {
            curData = data[i];
            if (this.numChars[curData]) {
                bucket += curData;
            }
            else {
                break;
            }
        }
        return { numPart: Number(bucket), data: data.substring(i) };
    };
    QVersionManager.prototype.getBigger = function (length1, length2) {
        return length1 > length2 ? length1 : length2;
    };
    QVersionManager.prototype.Add = function (versionData, versionData2) {
        return { data: versionData.data + versionData2.data, numPart: versionData.numPart + versionData2.numPart };
    };
    QVersionManager.prototype.ParseVersion = function (version) {
        var strings = version.split(".");
        var datas = [];
        var curData;
        for (var i = 0; i < 4; i++) {
            curData = strings.length > i ? this.extractPrependNumber(strings[i]) : { data: "", numPart: 0 };
            datas[i] = curData;
        }
        var ver = { major: datas[0], minor: datas[1], build: datas[2], revision: datas[3], length: strings.length };
        return ver;
    };
    QVersionManager.prototype.AddVersion = function (version, increment) {
        return {
            major: this.Add(version.major, increment.major),
            minor: this.Add(version.minor, increment.minor),
            revision: this.Add(version.revision, increment.revision),
            build: this.Add(version.build, increment.build),
            length: this.getBigger(version.length, increment.length),
        };
    };
    QVersionManager.prototype.IncrementAsQVersion = function (version, options) {
        var majorNumPart = options.major ? version.major.numPart + options.major : version.major.numPart;
        var minorNumPart = version.minor.numPart;
        var buildNumPart = version.build.numPart;
        var revisionNumPart = version.revision.numPart;
        if (!options.minor && !options.build && !options.revision && options.resetSubLevels == true) {
            minorNumPart = 0;
        }
        else if (options.minor) {
            minorNumPart = minorNumPart + options.minor;
        }
        if (!options.build && !options.revision && options.resetSubLevels == true) {
            buildNumPart = 0;
        }
        else if (options.build) {
            buildNumPart = buildNumPart + options.build;
        }
        if (!options.revision && options.resetSubLevels == true) {
            revisionNumPart = 0;
        }
        else if (options.revision) {
            revisionNumPart = revisionNumPart + options.revision;
        }
        return {
            major: { data: version.major.data, numPart: majorNumPart },
            minor: { data: version.minor.data, numPart: minorNumPart },
            build: { data: version.build.data, numPart: buildNumPart },
            revision: { data: version.revision.data, numPart: revisionNumPart },
            length: this.getBigger(version.length, options.revision ? 4 : options.build ? 3 : options.minor ? 2 : options.major ? 1 : 1),
        };
    };
    QVersionManager.prototype.ToString = function (version) {
        var keys = ["major", "minor", "build", "revision"];
        var v = version;
        var arr = arrayHelper_1.ArrayHelper.InitEmptyArray(version.length);
        for (var i = 0; i < version.length; i++) {
            var data = v[keys[i]];
            arr[i] = data.numPart + data.data;
        }
        return arr.join(".");
    };
    QVersionManager.prototype.IncrementAsString = function (version, options) {
        var parsedValue = this.ParseVersion(version);
        var incrementedValue = this.IncrementAsQVersion(parsedValue, options);
        var result = this.ToString(incrementedValue);
        return result;
    };
    return QVersionManager;
}());
exports.QVersionManager = QVersionManager;
