// @stechquick/algae/lib/helpers/windowHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WindowHelper = void 0;
var logger_1 = require("./logger");
var singularizer_1 = require("./singularizer");
var undefinedType = "undefined";
var WindowHelper = /** @class */ (function () {
    function WindowHelper() {
        this.doc = typeof (document) !== undefinedType ? document : {
            addEventListener: function (type, listener, options) {
                //might simulate this in future...
                logger_1.Logger.Log({ logMessage: "Enviroment does not support document. Triggering document.addEventListener now, type is: " + type, logType: logger_1.LogType.Warning });
                listener({ target: { readyState: this.readyState } });
            },
            createElement: function (tagName, options) {
                return { tagName: tagName, options: options };
            },
            readyState: "complete"
        };
        this.hist = typeof (history) !== undefinedType ? history : {
            pushState: function (data, title, url) {
                //might log data and url with additional data in logging
                logger_1.Logger.Log({ logMessage: "history.pushState not supported on Node. title: " + title, logType: logger_1.LogType.Debug });
            }
        };
        this.win = typeof (window) !== undefinedType ? window : {
            setInterval: setInterval,
            clearInterval: clearInterval,
            setTimeout: setTimeout,
            clearTimeout: clearTimeout,
            atob: atob,
            btoa: btoa
        };
        this.helperWorker = {
            worker: function (scriptURL, options) {
                return typeof (Worker) !== undefinedType ? new Worker(scriptURL, options) : undefined;
            }
        };
        // this.eventListeners = this.wrapEventListeners();
    }
    Object.defineProperty(WindowHelper.prototype, "onPopState", {
        set: function (value) {
            if (typeof (window) !== undefinedType) {
                window.onpopstate = value;
            }
            else {
                logger_1.Logger.Log({ logMessage: "onPopState not supported on Node.", logType: logger_1.LogType.Debug });
            }
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WindowHelper.prototype, "document", {
        get: function () {
            return this.doc;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WindowHelper.prototype, "history", {
        get: function () {
            return this.hist;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WindowHelper.prototype, "Window", {
        get: function () {
            return this.win;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(WindowHelper.prototype, "Worker", {
        get: function () {
            return this.helperWorker;
        },
        enumerable: false,
        configurable: true
    });
    WindowHelper.Instance = singularizer_1.Singularizer.trySingularize("theOneWindow", new WindowHelper().Window, new WindowHelper());
    return WindowHelper;
}());
exports.WindowHelper = WindowHelper;
