// @stechquick/symphony-cli/lib/predeploy/src/fn/fnPostdeploy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FnPostDeploy = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const predeployCommon_1 = require("../predeployCommon");
const log_1 = require("@stechquick/symphony-common/lib/log");
const chalk_1 = require("../../../helpers/chalk");
const fileHelper_1 = require("@stechquick/symphony-common/lib/helpers/fileHelper");
class FnPostDeploy {
    static execute() {
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Postdeploy starting..."), timeFormat: "datetime" });
        const deleteTargetItems = new Array();
        const options = {
            folder: "./functions",
            itemEndsWith: "func.yaml",
            iterator: (targetPath) => {
                log_1.Logger.log({ message: "Target path: " + targetPath });
                this.deleteList.forEach(deleteItem => {
                    log_1.Logger.log({ message: "Delete item: " + deleteItem });
                    const parsedDeleteItem = path.parse(deleteItem);
                    const targetItem = targetPath + parsedDeleteItem.name + parsedDeleteItem.ext;
                    deleteTargetItems.push(targetItem);
                });
            }
        };
        new predeployCommon_1.PredeployCommon().forEachFolderWithItem(options);
        this.deleteTargets(deleteTargetItems);
        log_1.Logger.log({ message: chalk_1.Chalk.create("green.bold", "Postdeploy finished successfully..."), timeFormat: "datetime" });
    }
    static deleteTargets(items) {
        items.forEach(delTarget => {
            if (!fs.existsSync(delTarget)) {
                return;
            }
            log_1.Logger.log({ message: "Deleting Item:" + delTarget });
            if (fileHelper_1.FileHelper.isDirectory(delTarget)) {
                fileHelper_1.FileHelper.removeFolder(delTarget);
                fs.rmdirSync(delTarget);
            }
            else {
                fs.unlinkSync(delTarget);
            }
        });
    }
}
exports.FnPostDeploy = FnPostDeploy;
FnPostDeploy.deleteList = [".npmrc", "package.json", "tsconfig.json", "Dockerfile", "package-lock.json", "node_modules", "sym-engine"];
