// @stechquick/symphony-cli/lib/helpers/microUIHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MicroUIHelper = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const cli_ux_1 = tslib_1.__importDefault(require("cli-ux"));
const fileHelper_1 = require("@stechquick/symphony-common/lib/helpers/fileHelper");
class MicroUIHelper {
    constructor() {
        this.subPath = "./";
        this.replaceMap = new Map();
        this.uiNameRegExp = /{{MicroUIName}}/g;
        this.uiDescRegExp = /{{MicroUIDesc}}/g;
        this.removeRegExp = /{{remove}}/g;
        this.createuiSearchPath = "/templates/createmicroui";
        this.replaceMap.set(this.removeRegExp, "");
    }
    async CreateMicroUI(microUIName, microUIDesc) {
        this.replaceMap.set(this.uiNameRegExp, microUIName);
        this.replaceMap.set(this.uiDescRegExp, microUIDesc);
        const existsSubPath = fs.existsSync(this.subPath + microUIName);
        if (existsSubPath) {
            const answer = await cli_ux_1.default.confirm("App: " + microUIName + " already exists. If you have any, you will lose your pending changes. Would you like to overwrite it?");
            if (!answer) {
                return;
            }
        }
        const templatePath = path.join(path.resolve(__dirname), "../.." + this.createuiSearchPath);
        fileHelper_1.FileHelper.exportFiles({ sourcePath: templatePath, targetPath: "./", replaceMap: this.replaceMap });
    }
}
exports.MicroUIHelper = MicroUIHelper;
