// @stechquick/symphony-cli/lib/helpers/packageHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PackageHelper = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const npmHelper_1 = require("./npmHelper");
const log_1 = require("@stechquick/symphony-common/lib/log");
const version_1 = require("@stechquick/algae/lib/helpers/version");
class PackageHelper {
    static getCliPackageJson() {
        const cliPath = PackageHelper.getCliPath();
        return PackageHelper.getPackageJson(cliPath + "package.json");
    }
    static getCliPath() {
        return path.join(path.resolve(__dirname), "../../");
    }
    static getBackwardPackages(settings) {
        const list = [];
        const cliFilePath = this.getCliPath();
        const packageJsonFilePath = path.join(cliFilePath, settings.packageJsonFilePath);
        const cliDevDeps = this.getPackageJson(packageJsonFilePath).devDependencies;
        const appDevDeps = npmHelper_1.NpmHelper.getCurrentPackageVersions();
        settings.packageNames.forEach(name => {
            log_1.Logger.log({ message: `checking package ${name}...`, level: log_1.LogLevel.log });
            const lastVersion = this.fixChars(cliDevDeps[name]);
            const usedVersion = appDevDeps[name];
            if (version_1.compareVersion(lastVersion, usedVersion) == 1) {
                list.push({ name, usedVersion, lastVersion });
            }
        });
        return list;
    }
    static getPackageJson(filePath) {
        const data = fs.readFileSync(path.resolve(filePath), { encoding: "utf8" });
        return JSON.parse(data);
    }
    static fixChars(value) {
        return value.replace(/[^0-9.]/g, "");
    }
}
exports.PackageHelper = PackageHelper;
