// json-schema-cycles/dist/index.js
import traverse from 'json-schema-traverse';
import { analyzeGraph, analyzeGraphFast, sortFullAnalysisResult, sortFastAnalysisResult, } from 'graph-cycles';
function decodeRef(ref) {
    if (ref == null)
        return undefined;
    if (ref.startsWith("#/definitions/"))
        return ref.slice(14);
    return undefined;
}
export function analyzeTypes(schema) {
    const graph = getJsonSchemaGraph(schema);
    return { ...analyzeGraph(graph), graph };
}
export function analyzeTypesFast(schema) {
    const graph = getJsonSchemaGraph(schema);
    return { ...analyzeGraphFast(graph), graph };
}
export function getJsonSchemaGraph(schema) {
    var _a;
    return Object
        .keys((_a = schema.definitions) !== null && _a !== void 0 ? _a : {})
        .map((type) => {
        const subSchema = schema.definitions[type];
        const dependencies = new Set();
        traverse(subSchema, (schema) => {
            const ref = decodeRef(schema.$ref);
            if (ref)
                dependencies.add(ref);
        });
        return [type, [...dependencies]];
    });
}
function sortGraph(result) {
    const { graph } = result;
    return [...graph]
        .sort((a, b) => a[0].localeCompare(b[0]))
        .map(([from, to]) => [from, [...to].sort()]);
}
export function sortTypeAnalysisFullResult(result) {
    return {
        ...sortFullAnalysisResult(result),
        graph: sortGraph(result),
    };
}
export function sortTypeAnalysisFastResult(result) {
    return {
        ...sortFastAnalysisResult(result),
        graph: sortGraph(result),
    };
}
