// typeconv/dist/convert-json-schema.js
import { convertJsonSchemaToCoreTypes, convertCoreTypesToJsonSchema, convertOpenApiToCoreTypes, convertCoreTypesToOpenApi, jsonSchemaDocumentToOpenApi, openApiToJsonSchema, } from "core-types-json-schema";
import { load as readYaml, dump as writeYaml } from "js-yaml";
import * as path from "path";
import { stringify } from "./utils.js";
import { userPackage, userPackageUrl } from "./package.js";
import { registerReader, registerWriter } from "./format-graph.js";
function maybeYamlReader(data, { warn, filename }) {
    const file = filename?.toLowerCase() ?? '';
    const isYaml = file.endsWith('yml') || file.endsWith('yaml');
    const input = isYaml
        ? readYaml(data, {
            filename,
            onWarning({ message, stack }) {
                warn(message, { blob: { stack } });
            },
        })
        // TODO: Maybe test if JSON parsing fails, and fallback to yaml,
        //       despite no filename match
        : data;
    return input;
}
export function getJsonSchemaReader() {
    return {
        kind: 'jsc',
        read(schema) {
            return convertJsonSchemaToCoreTypes(schema);
        },
    };
}
export function getJsonSchemaWriter() {
    return {
        kind: 'jsc',
        write(doc, { filename, sourceFilename }) {
            const { data: schemaObject, ...info } = convertCoreTypesToJsonSchema(doc, { filename, sourceFilename, userPackage, userPackageUrl });
            return {
                data: stringify(schemaObject),
                ...info,
            };
        },
        shortcut: {
            oapi(schema, readerOptions) {
                const openApiSchema = maybeYamlReader(schema, readerOptions);
                const jsonSchema = openApiToJsonSchema(openApiSchema);
                return {
                    data: stringify(jsonSchema),
                    convertedTypes: Object.keys(jsonSchema.definitions ?? {}),
                    notConvertedTypes: [],
                };
            }
        },
    };
}
export function getOpenApiReader() {
    return {
        kind: 'oapi',
        read(schema, readerOptions) {
            const openApiSchema = maybeYamlReader(schema, readerOptions);
            return convertOpenApiToCoreTypes(openApiSchema);
        },
    };
}
export function getOpenApiWriter({ format, ...openApiOptions }) {
    const formatOutput = (data) => (format === 'yaml' || format === 'yml')
        ? writeYaml(data)
        : stringify(data);
    const getOpenApiOptions = (filename) => ({
        ...openApiOptions,
        title: openApiOptions.title ??
            (filename
                ? `Converted from ${path.basename(filename)} with typeconv`
                : 'Converted with typeconv'),
        version: openApiOptions.version ?? '1',
    });
    return {
        kind: 'oapi',
        write(doc, { filename, sourceFilename }) {
            const { data: schemaObject, ...info } = convertCoreTypesToOpenApi(doc, {
                filename,
                sourceFilename,
                userPackage,
                userPackageUrl,
                ...getOpenApiOptions(filename),
            });
            return {
                data: formatOutput(schemaObject),
                ...info,
            };
        },
        shortcut: {
            jsc(data, { filename }) {
                const jsonSchema = JSON.parse(data);
                const openApiSchemaObject = jsonSchemaDocumentToOpenApi(jsonSchema, getOpenApiOptions(filename));
                return {
                    data: formatOutput(openApiSchemaObject),
                    convertedTypes: Object.keys(jsonSchema.definitions ?? {}),
                    notConvertedTypes: [],
                };
            }
        },
    };
}
const defaultOpenApiOptions = { format: 'json', title: '', version: '' };
registerReader(getJsonSchemaReader());
registerWriter(getJsonSchemaWriter());
registerReader(getOpenApiReader());
registerWriter(getOpenApiWriter(defaultOpenApiOptions));
