// @stechquick/symphony-cli/lib/helpers/namedComponent/ncHelper.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NCHelper = void 0;
const tslib_1 = require("tslib");
const fs = tslib_1.__importStar(require("fs"));
const path = tslib_1.__importStar(require("path"));
const cli_ux_1 = tslib_1.__importDefault(require("cli-ux"));
const fileHelper_1 = require("@stechquick/symphony-common/lib/helpers/fileHelper");
const log_1 = require("@stechquick/symphony-common/lib/log");
const npmHelper_1 = require("../npmHelper");
const chalk_1 = require("../../helpers/chalk");
class NCHelper {
    constructor() {
        this.subPath = "./";
        this.replaceMap = new Map();
        this.ncNameRegExp = /{{ProjectName}}/g;
        this.ncDescRegExp = /{{ProjectDescription}}/g;
        this.compNameRegExp = /{{ComponentName}}/g;
        this.removeRegExp = /{{remove}}/g;
        this.createncSearchPath = "../../../../templates/nc-create";
        this.addCompSearchPath = "../../../../templates/nc-addComp";
        this.replaceMap.set(this.removeRegExp, "");
    }
    static AssertInFolder() {
        return fs.existsSync("./src/namedComps") && fs.existsSync("./package.json");
    }
    async CreateNC(ncName, ncDesc) {
        this.replaceMap.set(this.ncNameRegExp, ncName);
        this.replaceMap.set(this.ncDescRegExp, ncDesc);
        const existsSubPath = fs.existsSync(this.subPath + ncName);
        if (existsSubPath) {
            const answer = await cli_ux_1.default.confirm("Named Component: " + ncName + " already exists. If you have any, you will lose your pending changes. Would you like to overwrite it?");
            if (!answer) {
                return;
            }
        }
        const templatePath = path.join(path.resolve(__dirname), "../.." + this.createncSearchPath);
        return fileHelper_1.FileHelper.exportFiles({ sourcePath: templatePath, targetPath: "./", replaceMap: this.replaceMap });
    }
    async AddComponent(componentName) {
        this.replaceMap.set(this.compNameRegExp, componentName);
        if (!NCHelper.AssertInFolder()) {
            log_1.Logger.log({ message: chalk_1.Chalk.create("red.bold", "Command must be executed in app root!") });
            return;
        }
        const existsSubPath = fs.existsSync("./src/namedComps" + componentName);
        if (existsSubPath) {
            const answer = await cli_ux_1.default.confirm("Component: " + componentName + " already exists. If you have any, you will lose your pending changes. Would you like to overwrite it?");
            if (!answer) {
                return;
            }
        }
        const templatePath = path.join(path.resolve(__dirname), "../.." + this.addCompSearchPath);
        fileHelper_1.FileHelper.exportFiles({ sourcePath: templatePath, targetPath: "./src/namedComps", replaceMap: this.replaceMap });
    }
    BuildNC(watch) {
        npmHelper_1.NpmHelper.buildWebpack({ watch });
    }
}
exports.NCHelper = NCHelper;
