// @stechquick/algae/lib/helpers/tsCompiler/compiler.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CommonTSCompiler = void 0;
var logger_1 = require("../logger");
var JSLoader_1 = require("./JSLoader");
var tsCompiler_1 = __importDefault(require("./tsCompiler"));
;
var CommonTSCompiler = /** @class */ (function () {
    function CommonTSCompiler() {
    }
    // private static dTSs: Record<string, ILangDTS> | undefined;
    CommonTSCompiler.LazyLoadTsCompiler = function (typescriptServicesURL) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                this.loadingProm = this.loadingProm || new Promise(function (res, rej) {
                    new JSLoader_1.JSLoader(typescriptServicesURL, function () { _this.loadingProm = undefined; res(); }, function (err) { _this.loadingProm = undefined; rej(err); }, 1).LoadJS();
                });
                return [2 /*return*/, this.loadingProm];
            });
        });
    };
    ;
    CommonTSCompiler.Compiler = function (options, compilerData, typescriptServicesURL, gmEditMode, targetVersion, moduleVersion) {
        return __awaiter(this, void 0, void 0, function () {
            var dtsProm, dtsList, errors, nameIndexedOptions, fss, compRes;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        dtsProm = this.LoadAndCacheDTSs(typescriptServicesURL);
                        if (!!window.ts) return [3 /*break*/, 2];
                        return [4 /*yield*/, this.LazyLoadTsCompiler(typescriptServicesURL)];
                    case 1:
                        _a.sent();
                        _a.label = 2;
                    case 2: return [4 /*yield*/, dtsProm];
                    case 3:
                        dtsList = _a.sent();
                        errors = Object.keys(dtsList).map(function (dtsName) { return dtsList[dtsName].error; }).filter(function (error) { return error; });
                        if (errors.length > 0) {
                            throw (errors);
                        }
                        nameIndexedOptions = options.reduce(function (prev, cur) { prev[cur.name] = cur; return prev; }, {});
                        fss = options.map(function (option) { return ({ name: option.name, body: _this.GenerateParamsIntelli(option.params) + "; " + option.script }); });
                        Object.keys(dtsList).forEach(function (dtsKey) { return fss.push({ name: dtsKey, body: dtsList[dtsKey].body }); });
                        Object.keys(compilerData).forEach(function (dtsKey) { return fss.push({ name: dtsKey, body: compilerData[dtsKey] }); });
                        compRes = (0, tsCompiler_1.default)(fss, (this.TypescriptCompilerOptions(window.ts, gmEditMode, targetVersion, moduleVersion)));
                        Object.keys(compRes).forEach(function (fileName) {
                            var cur = compRes[fileName];
                            if (!cur.diagnostic) {
                                return;
                            }
                            if (nameIndexedOptions[cur.name]) {
                                nameIndexedOptions[cur.name].errors = cur.diagnostic;
                            }
                        });
                        return [2 /*return*/, { options: options, compRes: compRes }];
                }
            });
        });
    };
    CommonTSCompiler.FetchModuleBody = function (url) {
        return new Promise(function (res, rej) { return fetch(url).then(function (response) { return response.text().then(function (body) { return res({ body: body, status: response.status }); }).catch(rej); }).catch(rej); });
    };
    CommonTSCompiler.LoadAndCacheDTSs = function (typescriptServicesURL) {
        return __awaiter(this, void 0, void 0, function () {
            var items, leftItems, proms, files;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        typescriptServicesURL = typescriptServicesURL.replace(/\/[\d.]+\/typescriptServices\.js$/, "");
                        items = [typescriptServicesURL + "/lib/lib.es5.d.ts", typescriptServicesURL + "/lib/lib.es2015.symbol.d.ts", typescriptServicesURL + "/lib/lib.es2015.symbol.wellknown.d.ts", typescriptServicesURL + "/lib/lib.es2020.bigint.d.ts", typescriptServicesURL + "/lib/lib.es2015.promise.d.ts", typescriptServicesURL + "/lib/lib.es2017.string.d.ts", typescriptServicesURL + "/lib/lib.es2015.core.d.ts"];
                        leftItems = items.filter(function (item) { var _a; return !_this.loadedDTSs[item] || ((_a = _this.loadedDTSs[item]) === null || _a === void 0 ? void 0 : _a.error); });
                        if (leftItems.length <= 0) {
                            return [2 /*return*/, this.loadedDTSs];
                        }
                        proms = leftItems.map(function (item) { return _this.FetchModuleBody(item); });
                        return [4 /*yield*/, Promise.all(proms)];
                    case 1:
                        files = _a.sent();
                        files.forEach(function (file, i) {
                            var _a;
                            _this.loadedDTSs[leftItems[i]] = (file.status == 200 && file.body) ? { body: file.body } : { error: file.status.toString() };
                            if (file.status != 200) {
                                logger_1.Logger.log({ message: "file load failed " + leftItems[i] + ": " + file.status, error: (_a = _this.loadedDTSs[leftItems[i]]) === null || _a === void 0 ? void 0 : _a.error, level: logger_1.LogLevel.error });
                            }
                            ;
                        });
                        return [2 /*return*/, this.loadedDTSs];
                }
            });
        });
    };
    CommonTSCompiler.GenerateParamsIntelli = function (params) {
        var _this = this;
        if (!params) {
            return "";
        }
        if (!(params instanceof Array)) {
            return params.split(",").map(function (param) { return "declare var " + param; }).join(";\n");
        }
        return params.map(function (param) { return "declare var " + param.Name + ": " + _this.PropTypeToTs(param.Type) + ";"; }).join("\n");
    };
    CommonTSCompiler.PropTypeToTs = function (type, usage) {
        var _a;
        if (usage === void 0) { usage = "prop"; }
        var lookup = {};
        lookup[Boolean] = "boolean";
        lookup[String] = "string";
        lookup[Number] = "number";
        lookup[Object] = "object";
        lookup[Array] = "Array<any>";
        lookup[Function] = "Function";
        lookup[undefined] = "any";
        lookup[null] = "any";
        if (usage == "retVal") {
            lookup[undefined] = "void";
            lookup[null] = "void";
        }
        return (_a = lookup[type]) !== null && _a !== void 0 ? _a : type;
    };
    CommonTSCompiler.TypescriptCompilerOptions = function (tsInst, gmEditMode, targetVersion, moduleVersion) {
        var target = tsInst ? tsInst.ScriptTarget[targetVersion || 'ES5'] : 1;
        var module = tsInst && moduleVersion ? tsInst.ModuleKind[moduleVersion] : undefined;
        var x = {
            noLib: true,
            target: target,
            module: module,
            allowNonTsExtensions: true,
            strict: false,
            noImplicitUseStrict: true,
            inlineSources: gmEditMode,
            inlineSourceMap: gmEditMode
        };
        return x;
    };
    ;
    CommonTSCompiler.loadedDTSs = {};
    return CommonTSCompiler;
}());
exports.CommonTSCompiler = CommonTSCompiler;
