// @stechquick/algae/lib/helpers/tsCompiler/JSLoader.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSLoader = void 0;
var JSLoader = /** @class */ (function () {
    function JSLoader(jsName, successCb, failCb, tryCount, noCache, cacheVersion, options) {
        if (noCache === void 0) { noCache = false; }
        if (cacheVersion === void 0) { cacheVersion = 0; }
        if (options === void 0) { options = {}; }
        this.JSName = jsName;
        this.SuccessCB = successCb;
        this.FailCB = failCb;
        this.TryCount = tryCount;
        this.noCache = noCache;
        this.cacheVersion = cacheVersion;
        this.options = options;
    }
    JSLoader.prototype.LoadJS = function (type) {
        var cbObject = this.createCallBackObject(this.JSName, this.SuccessCB, this.FailCB, this.TryCount, this.noCache, this.cacheVersion, this.options);
        cbObject.goLoad(type);
    };
    JSLoader.prototype.createCallBackObject = function (jsName, successCb, failCb, tryCount, noCache, cacheVersion, options) {
        var retryLeft = tryCount;
        var scriptElem;
        var isJSLoaded = false;
        var interval = 100;
        var isBrowser = !!(typeof window !== 'undefined' && window.document);
        var readyRegExp = isBrowser ? /^complete$/ : /^(complete|loaded)$/;
        var cbObject = {
            retryLogic: function (isLoaded) {
                isJSLoaded = isLoaded;
                retryLeft--;
                if (isJSLoaded || retryLeft < 1) {
                    return;
                }
                scriptElem.parentElement.removeChild(scriptElem);
                interval *= 3;
                setTimeout(function () { cbObject.goLoad(); }, interval);
            },
            onScriptLoad: function (evt) {
                cbObject.retryLogic(true);
                var node = evt.currentTarget || evt.srcElement;
                if (evt.type === 'load' || (readyRegExp.test(node.readyState))) {
                    cbObject.clearListeners(evt);
                    if (successCb) {
                        successCb();
                    }
                }
            },
            onScriptError: function (evt) {
                cbObject.clearListeners(evt);
                cbObject.retryLogic(false);
                if (failCb) {
                    failCb(evt);
                }
            },
            addListeners: function () {
                scriptElem.addEventListener('load', cbObject.onScriptLoad, false);
                scriptElem.addEventListener('error', cbObject.onScriptError, false);
            },
            clearListeners: function (evt) {
                var node = evt.currentTarget || evt.srcElement;
                node.removeEventListener('load', cbObject.onScriptLoad, false);
                node.removeEventListener('error', cbObject.onScriptError, false);
            },
            goLoad: function (type) {
                scriptElem = document.createElement('script');
                cbObject.addListeners();
                scriptElem.src = noCache ? jsName + "?v=" + cacheVersion : jsName;
                ;
                scriptElem.type = type || "text/javascript";
                cbObject.setTagProps();
                if (options.charset) {
                    scriptElem.charset = options.charset;
                }
                document.head.appendChild(scriptElem);
            },
            setTagProps: function () {
                if (!options.scriptTagProps) {
                    return;
                }
                var props = options.scriptTagProps;
                Object.keys(props).reduce(function (prev, curKey) { prev[curKey] = props[curKey]; return prev; }, scriptElem);
            }
        };
        return cbObject;
    };
    ;
    return JSLoader;
}());
exports.JSLoader = JSLoader;
