// @stechquick/algae/lib/helpers/tsCompiler/tsCompiler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var ts = __importStar(require("typescript"));
;
;
var extTs = function (tsInst) { return tsInst; };
var Compile = function (files, options) {
    var putFileToFolder = function (rootFolder, path, file) {
        var folders = path.split("/");
        var fileName = folders.splice(folders.length - 1)[0];
        var lastFolder = folders.reduce(function (prev, cur) { return prev[cur] = {}; }, rootFolder);
        lastFolder[fileName] = file;
    };
    var compiledFiles = files.reduce(function (prev, cur) { putFileToFolder(prev, cur.name, { name: cur.name, body: cur.body }); return prev; }, {});
    var fileNames = files.map(function (file) { return file.name; });
    var fileSources = {};
    var ext = extTs(ts);
    files.forEach(function (cur) {
        fileSources[ext.normalizePath(cur.name)] = ts.createSourceFile(cur.name, cur.body, ts.ScriptTarget.ES5);
    });
    var newLine = ext.getNewLineCharacter(options);
    var compilerHost = {
        getSourceFile: function (fileName) { return fileSources[fileName]; },
        writeFile: function (name, text) {
            var extIndex = name.indexOf(".js");
            var fileNameWithExt = name.substring(0, extIndex);
            var tsName = fileNameWithExt.endsWith('.ts') ? fileNameWithExt : fileNameWithExt + ".ts";
            if (ext.fileExtensionIs(name, ".map")) {
                compiledFiles[tsName].sourceMapText = text;
            }
            else {
                compiledFiles[tsName].outputText = text;
            }
        },
        getDefaultLibFileName: function () { return "lib.d.ts"; },
        useCaseSensitiveFileNames: function () { return false; },
        getCanonicalFileName: function (fileName) { return fileName; },
        getCurrentDirectory: function () { return ""; },
        getNewLine: function () { return newLine; },
        fileExists: function (fileName) { return files[fileName] != undefined; },
        readFile: function () { return ""; },
        directoryExists: function (dir) { return dir.indexOf("type") <= -1; },
        getDirectories: function () { return []; }
    };
    options.noLib = true;
    options.inlineSourceMap = false;
    var program = ts.createProgram(fileNames, options, compilerHost);
    var emitResult = program.emit();
    var allDiagnostics = ts.getPreEmitDiagnostics(program).concat(emitResult.diagnostics);
    allDiagnostics.forEach(function (diagnostic) {
        if (diagnostic.file) {
            var _a = diagnostic.file.getLineAndCharacterOfPosition(diagnostic.start), line = _a.line, character = _a.character;
            var message = ts.flattenDiagnosticMessageText(diagnostic.messageText, "\n");
            var curCompiledFile = compiledFiles[diagnostic.file.fileName] || {};
            curCompiledFile.diagnostic = curCompiledFile.diagnostic || [];
            curCompiledFile.diagnostic.push(diagnostic.file.fileName + " (" + (line + 1) + "," + (character + 1) + "): " + message);
        }
    });
    return compiledFiles;
};
exports.default = Compile;
