// @stechquick/symphony-cli/templates/init/{{AppName}}/functions/samplefunc/samplefunc.ts
import { ISymphonyObject, symphony } from "@stechquick/symphony";
import { TReqDefault, TReqHeaderDefault } from "@stechquick/symphony/dist/src/infrastructure/managers/symphonyObject/symphonyRequest";
import { TRespDefault, TRespHeaderDefault } from "@stechquick/symphony/dist/src/infrastructure/managers/symphonyObject/symphonyResponse";

symphony.registerFunc({
    funcName: "samplefunc",
    keepBag: false,
    requireFunc: require,
    func: samplefunc,
});

export interface IRequest extends TReqDefault { };
export interface IRequestHeader extends TReqHeaderDefault { };
export interface IResponse extends TRespDefault { };
export interface IResponseHeader extends TRespHeaderDefault { };

async function samplefunc(sym: ISymphonyObject<IRequest, IRequestHeader, IResponse, IResponseHeader>) {
    const x = sym.request.body.number1;
    const y = sym.request.body.number2;
    const sum: number = x + y;
    const difference: number = y - x;
    sym.response.body.calculations = { funcName: "samplefunc", sum: sum, difference: difference };

    //#region rest example
    const respGet = await sym.rest.get({ symConfigKey: "restGetExample" });
    sym.response.body.restGetResponse = { listEmp: respGet.body };
    //#endregion

    //#region rest example
    const respPost = await sym.rest.post({ symConfigKey: "restPostExample", body: { "name": "test", "salary": "123", "age": "23" } });
    sym.response.body.restPostResponse = { listEmp: respPost.body };
    //#endregion

    //#region soap example
    const soapBody = { "intA": "2", "intB": "3" };
    const respSoap = await sym.soap.request({ request: { body: soapBody, methodName: "Add", symConfigKey: "soapExample" } });
    sym.response.body.soapResponse = { total: respSoap.data };
    //#endregion
}
