// @stechquick/symphony-nanoserver/dist/src/RouteParser.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RouteParser = void 0;
var path_1 = __importDefault(require("path"));
var url_1 = __importDefault(require("url"));
;
;
var RouteParser = /** @class */ (function () {
    function RouteParser(templatePath) {
        var _this = this;
        var pathParts = templatePath.split("/").filter(function (pathPart) { return pathPart; });
        this.staticParts = Array(pathParts.length);
        this.dynamicParts = Array(pathParts.length);
        var counter = 0;
        pathParts.forEach(function (pathPart) {
            if (pathPart.startsWith("{")) {
                _this.dynamicParts[counter] = pathPart.substring(1, pathPart.length - 1);
            }
            else {
                _this.staticParts[counter] = pathPart.toLowerCase();
            }
            counter++;
        });
    }
    RouteParser.ParseUrl = function (urlString) {
        var _a;
        var urlData = url_1.default.parse(urlString);
        var parsedPath = path_1.default.parse(urlData.pathname);
        var args = (_a = urlData.query) !== null && _a !== void 0 ? _a : "";
        return __assign(__assign({}, parsedPath), { args: args });
    };
    RouteParser.prototype.MatchAndGetRouteDict = function (parsedUrl) {
        var retVal = { match: false, routeDict: {}, power: 0 };
        var pureUrl = parsedUrl.dir + "/" + parsedUrl.name + parsedUrl.ext;
        var incommingPathParts = pureUrl.split("/").filter(function (part) { return part; });
        if (incommingPathParts.length < this.staticParts.length) {
            return retVal;
        }
        var counter = 0;
        for (var i = 0, len = this.staticParts.length; i < len; i++) {
            var dynamicPart = this.dynamicParts[i];
            var currentStatic = this.staticParts[i];
            retVal.match = true;
            if (currentStatic == "*") {
                break;
            }
            if (dynamicPart) {
                retVal.routeDict[dynamicPart] = decodeURIComponent(incommingPathParts[i]);
            }
            else if (incommingPathParts[counter].length > 0) {
                if (incommingPathParts[counter].toLowerCase() != currentStatic) {
                    retVal.match = false;
                    break;
                }
            }
            counter++;
        }
        if (retVal.match) {
            retVal.getMatchData = function () { return ({
                args: parsedUrl.args,
                matchLeftOver: incommingPathParts.length >= counter ? incommingPathParts.slice(counter).map(function (item) { return "/" + item; }).join("") : undefined,
            }); };
            retVal.power = counter;
        }
        return retVal;
    };
    return RouteParser;
}());
exports.RouteParser = RouteParser;
