// @stechquick/algae/lib/helpers/socket/domSocket.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DomSocket = exports.DomSocketStatus = exports.ConnectReason = exports.ReconnectType = void 0;
var logger_1 = require("../logger");
var ReconnectType;
(function (ReconnectType) {
    ReconnectType[ReconnectType["BackOff"] = 0] = "BackOff";
    ReconnectType[ReconnectType["Urgent"] = 1] = "Urgent";
})(ReconnectType = exports.ReconnectType || (exports.ReconnectType = {}));
var ConnectReason;
(function (ConnectReason) {
    ConnectReason[ConnectReason["ConnectCall"] = 0] = "ConnectCall";
    ConnectReason[ConnectReason["SendReconnect"] = 1] = "SendReconnect";
    ConnectReason[ConnectReason["ErrorReconnect"] = 2] = "ErrorReconnect";
    ConnectReason[ConnectReason["CloseReconnect"] = 3] = "CloseReconnect";
})(ConnectReason = exports.ConnectReason || (exports.ConnectReason = {}));
var DomSocketStatus;
(function (DomSocketStatus) {
    DomSocketStatus[DomSocketStatus["Initial"] = 0] = "Initial";
    DomSocketStatus[DomSocketStatus["Disconnected"] = 1] = "Disconnected";
    DomSocketStatus[DomSocketStatus["WaitingToConnect"] = 2] = "WaitingToConnect";
    DomSocketStatus[DomSocketStatus["Connecting"] = 3] = "Connecting";
    DomSocketStatus[DomSocketStatus["Connected"] = 4] = "Connected";
})(DomSocketStatus = exports.DomSocketStatus || (exports.DomSocketStatus = {}));
var DomSocket = /** @class */ (function () {
    function DomSocket(url, options) {
        if (options === void 0) { options = {}; }
        this._status = DomSocketStatus.Initial;
        this.msgSenderList = [];
        this.seqErrCount = 0;
        this.reconnectTryCount = 0;
        this.timerID = undefined;
        this.url = url;
        this.id = url.length > 40 ? url.substring(0, 40) : url;
        this.options = options;
        this.options.minLogLevel = this.options.minLogLevel || logger_1.LogLevel.log;
    }
    Object.defineProperty(DomSocket.prototype, "status", {
        get: function () {
            return this._status;
        },
        set: function (value) {
            var _a, _b;
            if (value === this._status) {
                return;
            }
            this._status = value;
            (_b = (_a = this.options).onSocketStatus) === null || _b === void 0 ? void 0 : _b.call(_a, this._status);
        },
        enumerable: false,
        configurable: true
    });
    DomSocket.prototype.destroy = function () {
        this.Close();
    };
    Object.defineProperty(DomSocket.prototype, "OnConnect", {
        set: function (cb) {
            this.options.onConnect = cb;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomSocket.prototype, "OnMessage", {
        set: function (cb) {
            this.options.onMessage = cb;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomSocket.prototype, "ReConnect", {
        set: function (option) {
            this.options.reConnect = option;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomSocket.prototype, "MinLogLevel", {
        set: function (minLogLevel) {
            this.options.minLogLevel = minLogLevel;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomSocket.prototype, "Status", {
        get: function () {
            return this.status;
        },
        enumerable: false,
        configurable: true
    });
    Object.defineProperty(DomSocket.prototype, "onSocketStatus", {
        set: function (cb) {
            this.options.onSocketStatus = cb;
        },
        enumerable: false,
        configurable: true
    });
    DomSocket.prototype.Send = function (message) {
        return __awaiter(this, void 0, void 0, function () {
            var data;
            var _this = this;
            return __generator(this, function (_a) {
                data = message instanceof Object ? JSON.stringify(message) : message;
                return [2 /*return*/, new Promise(function (res, rej) {
                        var msgSender = function () {
                            _this.log({ message: "Sending msg: " + data, level: logger_1.LogLevel.debug });
                            _this.ws.send(data);
                            res(true);
                        };
                        var reason = _this.reConnect(ConnectReason.SendReconnect, msgSender);
                        if (reason) {
                            res(reason);
                        }
                    })];
            });
        });
    };
    /**
     *
     * @param code 4200-4300. -> https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent#Status_codes
     * @param reason
     */
    DomSocket.prototype.Close = function (code, reason) {
        var ws = this.ws;
        this.ws = undefined;
        if (!ws) {
            return;
        }
        var userCloded = DomSocket.CodeAndReasons.userCloded;
        code = code !== null && code !== void 0 ? code : userCloded.code;
        reason = reason !== null && reason !== void 0 ? reason : userCloded.reason;
        this.log({ message: "Close call code: " + code + " reason: " + reason, level: logger_1.LogLevel.debug });
        this.options.reConnect = undefined;
        ws.close(code, reason);
    };
    DomSocket.prototype.Connect = function () {
        this.reConnect(ConnectReason.ConnectCall);
        return this;
    };
    DomSocket.prototype.log = function (options) {
        var _a, _b;
        if (this.options.minLogLevel && logger_1.LogLevelLookup[options.level] < logger_1.LogLevelLookup[this.options.minLogLevel]) {
            return;
        }
        var message = "Client " + this.id + ": " + options.message;
        (_b = (_a = this.options).logCallBack) === null || _b === void 0 ? void 0 : _b.call(_a, { message: message, additional: options.additional, level: options.level, timeFormat: "time" });
    };
    DomSocket.prototype.reConnect = function (reason, msgSender) {
        var _this = this;
        var _a;
        switch ((_a = this.ws) === null || _a === void 0 ? void 0 : _a.readyState) {
            case WebSocket.OPEN:
                msgSender === null || msgSender === void 0 ? void 0 : msgSender();
                return; // if open, send and return
            case WebSocket.CONNECTING:
                msgSender && this.msgSenderList.push(msgSender);
                return; // if connecting, queue and return
            case WebSocket.CLOSED:
            case WebSocket.CLOSING:
            default: // ws -> null
                break;
        }
        if (reason == ConnectReason.ConnectCall) {
            msgSender && this.msgSenderList.push(msgSender); // queue and try connection
            this.connect(reason);
            return;
        }
        if (!this.options.reConnect) {
            return "Closed";
        } // no reconnect, fail and return
        msgSender && this.msgSenderList.push(msgSender); // queue and try re-connection
        if (this.timerID) {
            return;
        } // it is allready timered
        this.status = DomSocketStatus.WaitingToConnect;
        this.reconnectTryCount++;
        var timeoutCounter = this.calculateReconnectInterval(this.options.reConnect);
        this.timerID = window.setTimeout(function () {
            _this.timerID = undefined;
            _this.connect(reason);
        }, timeoutCounter);
    };
    DomSocket.prototype.connect = function (reason) {
        var _this = this;
        this.status = DomSocketStatus.Connecting;
        this.ws = new WebSocket(this.url);
        this.ws.onclose = this.onSocketClose.bind(this);
        this.ws.onopen = function (openEvent) { return _this.onSocketOpen(openEvent, reason); };
        this.ws.onerror = this.onSocketError.bind(this);
        this.ws.onmessage = this.onSocketMessage.bind(this);
    };
    DomSocket.prototype.onSocketError = function (errorEvent) {
        var reconMsg = "";
        if ((!this.ws || this.ws.readyState == WebSocket.CLOSED) && this.options.reConnect) {
            this.seqErrCount++;
            reconMsg = this.options.reConnect ? ", errCount: " + this.seqErrCount : "";
            this.reConnect(ConnectReason.ErrorReconnect);
        }
        this.log({ message: "error" + reconMsg, level: logger_1.LogLevel.error, additional: errorEvent });
    };
    DomSocket.prototype.calculateReconnectInterval = function (options) {
        var calculatedInterval = options.intervalMs;
        if (options.type == ReconnectType.BackOff) {
            var fibonacciArray = [1, 2, 3, 5, 8, 13, 21, 34, 55];
            var intervalUnit = options.intervalMs / 5;
            var fibonacciRatio = this.reconnectTryCount > fibonacciArray.length ? fibonacciArray[fibonacciArray.length - 1] : fibonacciArray[this.reconnectTryCount - 1];
            calculatedInterval = Math.round(intervalUnit * fibonacciRatio);
        }
        return calculatedInterval;
    };
    DomSocket.prototype.onSocketOpen = function (openEvent, reason) {
        var _a, _b, _c;
        this.status = DomSocketStatus.Connected;
        this.log({ message: "connected", level: logger_1.LogLevel.log });
        (_b = (_a = this.options).onConnect) === null || _b === void 0 ? void 0 : _b.call(_a, reason);
        this.seqErrCount = 0;
        this.reconnectTryCount = 0;
        while (this.msgSenderList.length) {
            (_c = this.msgSenderList.pop()) === null || _c === void 0 ? void 0 : _c();
        }
    };
    DomSocket.prototype.onSocketClose = function (closeEvent) {
        this.status = DomSocketStatus.Disconnected;
        this.log({ message: "error", level: logger_1.LogLevel.log });
        var userCloded = DomSocket.CodeAndReasons.userCloded;
        if (closeEvent.code == userCloded.code && closeEvent.reason == userCloded.reason) {
            return;
        }
        if (!this.options.reConnect) {
            return;
        }
        this.reConnect(ConnectReason.CloseReconnect);
    };
    DomSocket.prototype.onSocketMessage = function (messageEvent) {
        var _a, _b;
        this.log({ message: "message", level: logger_1.LogLevel.debug });
        (_b = (_a = this.options).onMessage) === null || _b === void 0 ? void 0 : _b.call(_a, messageEvent.data);
    };
    //https://developer.mozilla.org/en-US/docs/Web/API/CloseEvent#Status_codes
    DomSocket.CodeAndReasons = { userCloded: { code: 4000, reason: "user closed" } };
    return DomSocket;
}());
exports.DomSocket = DomSocket;
