// @stechquick/symphony-cli/templates/createmicroui/{{MicroUIName}}-ui/webpack.config.js
const path = require('path');
const VueLoaderPlugin = require('vue-loader/lib/plugin');
const WebpackBar = require('webpackbar');
const moduleName = require('./indexModuleName');
const moduleFileName = `${moduleName}.js`;
let domainCompCount=0;
let domainComponentList = {} 
try {
    const compNames = require('./indexVCompNames');    
    domainCompCount= Object.keys(compNames).length
    if(domainCompCount>0){
        domainComponentList = Object.keys(compNames).reduce((prev, compName) => { prev[compName + ".js"] = compNames[compName].file; return prev; }, domainComponentList);
    }

    
} catch (error) {
    console.log("indexVCompNames dosyasını valid hale getiriniz.")
}


const components = {
    // entry: './indexInitial.js',
    entry: domainComponentList,
    devtool: 'source-map',
    mode: 'development',
    watch: false,
    resolve: {
        extensions: ['.vue', '.js'],
    },
    output: {
        // filename: moduleFileName,
        filename: '[name]',
        path: path.resolve(__dirname, 'dist'),
        // chunkFilename: '[name].main.js',
        // publicPath: 'dist/'
        libraryTarget: 'jsonp',
        library: `vuetify_microUI.registerModule("${moduleFileName}").registerComponent`
    },
    module: {
        rules: [
            {
                test: /\.vue$/,
                loader: 'vue-loader'
            },
            {
                test: /\.tsx?$/,
                loader: "ts-loader",
                exclude: /node_modules|vue\/src/,
                options: {
                    appendTsSuffixTo: [/\.vue$/],
                }
            },
            {
                test: /\.css$/,
                use: [{
                    loader: 'style-loader'
                }, {
                    loader: 'css-loader'
                }],
            },
            {
                test: /\.styl$/,
                use: [
                    'style-loader',
                    'css-loader',
                    {
                        loader: 'stylus-loader',
                        options: {
                            use: [],
                        },
                    },
                ],
            },
            {
                test: /\.js$/,
                use: ["source-map-loader"],
                enforce: "pre"
            },
            {
                test: /\.js$/,
                loader: "babel-loader",
                options: {
                    presets: ["@babel/preset-env"],
                    babelrc: false,
                    cacheDirectory: false,
                    compact: false
                },
                include: [
                    path.resolve(__dirname, './src'),
                    path.resolve(__dirname, './index'),
                    path.resolve(__dirname, './indexInitial'),

                ],
            },
            {
                test: /\.(woff2?|eot|ttf|png|jpg|svg|otf)(\?.*)?$/,
                loader: 'url-loader',
            },
            {
                test: /\.s(a|c)ss$/,
                // test: /\.sass$/,
                loaders: ['style-loader', 'css-loader', 'sass-loader']
            }
        ]
    },
    plugins: [new VueLoaderPlugin(), new WebpackBar()]

};

const indexConfig = {
    entry: './indexInitial.js',
    //entry: domainComponentList,
    devtool: 'source-map',
    mode: 'development',
    watch: false,
    resolve: {
        extensions: ['.vue', '.js'],
    },
    output: {
        filename: moduleFileName,
        // filename: '[name]',
        path: path.resolve(__dirname, 'dist'),
        // chunkFilename: '[name].main.js',
        // publicPath: 'dist/'
        libraryTarget: 'jsonp',
        library: `rally_microUI.registerModule("${moduleFileName}").registerInitial`
    },
    module: {
        rules: [
            {
                test: /\.vue$/,
                loader: 'vue-loader'
            },
            {
                test: /\.tsx?$/,
                loader: "ts-loader",
                exclude: /node_modules|vue\/src/,
                options: {
                    appendTsSuffixTo: [/\.vue$/],
                }
            },
            {
                test: /\.css$/,
                use: [{
                    loader: 'style-loader'
                }, {
                    loader: 'css-loader'
                }],
            },
            {
                test: /\.styl$/,
                use: [
                    'style-loader',
                    'css-loader',
                    {
                        loader: 'stylus-loader',
                        options: {
                            use: [],
                        },
                    },
                ],
            },
            {
                test: /\.js$/,
                use: ["source-map-loader"],
                enforce: "pre"
            },
            {
                test: /\.js$/,
                loader: "babel-loader",
                options: {
                    presets: ["@babel/preset-env"],
                 
                    babelrc: false,
                    cacheDirectory: false,
                    compact: false
                },
                include: [
                    path.resolve(__dirname, './src'),
                    path.resolve(__dirname, './index'),
                    path.resolve(__dirname, './indexInitial'),


                    // path.appSrc,
                    // path.appNodeModules
                ],
            },
            {
                test: /\.(woff2?|eot|ttf|png|jpg|svg|otf)(\?.*)?$/,
                loader: 'url-loader',
            },
            {
                test: /\.s(a|c)ss$/,
                // test: /\.sass$/,
                loaders: ['style-loader', 'css-loader', 'sass-loader']
            }
        ]
    },
    plugins: [new VueLoaderPlugin(), new WebpackBar()]

};


const compNamesConfig = {
    entry: './indexCompNames.js',
    mode: 'development',
    watch: false,
    module: {
        rules: [
            {
                test: /\.js$/,
                use: {
                    loader: 'babel-loader',
                },
                exclude: [path.resolve(__dirname, './node_modules')]
            }
        ]
    },
    plugins: [new WebpackBar()],
    output: {
        filename: "compNames_" + moduleFileName,
        path: path.resolve(__dirname, 'dist'),
        libraryTarget: 'jsonp',
        library: `rally_microUI.registerModule("${moduleFileName}").registerCompNames`
    }
};


const compVNamesConfig = {
    entry: './indexVCompNames.json',
    mode: 'development',
    watch: false,
    module: {
        rules: [
            {
                test: /\.js$/,
                use: {
                    loader: 'babel-loader',
                },
                exclude: [path.resolve(__dirname, './node_modules')]
            }
        ]
    },
    plugins: [new WebpackBar()],
    output: {
        filename: "compVNames_" + moduleFileName,
        path: path.resolve(__dirname, 'dist'),
        libraryTarget: 'jsonp',
        library: `vuetify_microUI.registerModule("${moduleFileName}").registerCompNames`
    }
};
const exportedList=[indexConfig, compNamesConfig];
if(domainCompCount>0 ){
    exportedList.push(components,compVNamesConfig);
}

module.exports= exportedList;
